Project-wide LLM policy (#3959)
rust-lang/rfcs: text/3959-llm-policy.md

Reviewer: I love the detailed write-up on the concerns and effectiveness of GenAI, well done! <3

I'll add my own views to the discussion.

The focus on honesty is something I can respect, but unfortunately being honest about doing something unethical does not make it okay. It should not excuse the act. If I steal your bag of chips, and then state I stole it, it does not mean I get away with it. That's just an admission of guilt.

It would be a very bad look for the Rust project if it knowingly accepted contributions authored by GenAI, and would be very sad. The Rust project has from the start made ethical stances, such as focusing on being inclusive. This should not be different.

I would also explicitly add that the policy covers GenAI images, video, and everything else. The rules for code and other GenAI should not be different. In my views, we should be in solidarity with other affected professions in standing against it.

As a community, we should not look the other way when faced with the massive ethical issues and injustice with GenAI. We should not accept it. No-one should be able to proudly say they've used it for a contribution and then get their contribution accepted.

As for usage of GenAI in research, prototyping, search and similar cases, where the output is not part of the contribution, I feel it should be discouraged, but we can't realistically do more than that. With authorship, usage of GenAI should be banned completely.

Accessibility-wise, I am fine with a carve-out, as long as it is not too broad. The situation here is unfortunate; if all the money wasted on GenAI was instead spent on accessibility then we would have way better accessibility and it would be ethical. Sadly, that is not the world we live in, so we need to make do with what we have.

Also, I believe the policy should be a minimum baseline. Projects should be able to make it stricter, but not weaker.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Reviewer 2: > If I steal your bag of chips, and then state I stole it, it does not mean I get away with it. That's just an admission of guilt.

Nothing in this RFC states that you get away from it or not.

Reviewer: > > If I steal your bag of chips, and then state I stole it, it does not mean I get away with it. That's just an admission of guilt.
> 
> Nothing in this RFC states that you get away from it or not.

I don't know what you mean by this. It states that disclosed nontrivial GenAI use is okay. (and in case of trivial, even undisclosed)

Author: FWIW, part of the goal of this comment:

> To keep things focused on policy, there are two broad categories of comments we'd like to request you avoid:
>
> * Simply stating your viewpoint on LLMs, even if you provide reasons. While these arguments can be useful for the RFC, they are better worded as explicit suggestions to specific areas of the RFC, rather than as just general comments.

Is mostly to avoid situations where things devolve into a discussion on whether ethics ultimately make usage irredeemable or not, since I don't think that it's going to go particularly well on the RFC. For what it's worth, I agree with you, although reiterating what I mentioned on fedi, the goal is to:

1. Push the bar in a direction that focuses on accountability (if you do use an LLM, be honest about it and don't make the code worse for it)
2. Put the ethical arguments on the table so that they can be used later to potentially strengthen the policy to reduce LLM usage

Right now, I'm only aiming for a low bar mostly because I feel like it's most likely to be accepted, and right now there's kind of a lot of open questions that are causing headaches for reviewers. Having *a* policy in place will make it easier to change the policy, etc.

Also, FWIW, I do explicitly outline in the policy that usage is not encouraged, only allowed for pragmatic reasons.

----

To respond to a few other points of feedback:

> I would also explicitly add that the policy covers GenAI images, video, and everything else. The rules for code and other GenAI should not be different. In my views, we should be in solidarity with other affected professions in standing against it.

So, right now the policy *does* clarify that both text and code are included, although I didn't mention images, video, and audio since it's not particularly relevant here. However, I do think that this is kind of a weak argument and it's good to be specific.

> Accessibility-wise, I am fine with a carve-out, as long as it is not too broad.

This is kind of what I aimed to go for with the trivial usage definition; the goal is that if the end result was just accessibility, or something that otherwise is indistinguishable, we probably shouldn't be wasting enforcement bandwidth on trying to litigate that. Additionally, it pushes people to understand what they're really using LLMs for, which hopefully will lead to people understanding their effects better.

Wikipedia's policy is one mentioned in the prior art that stands out pretty well: they allow LLMs for copy-editing but my understanding is that this is mostly pragmatic for enforcement reasons, and they can't really preserve a tone that condones LLM usage when things are publicly editable. Things are different here since you can't just edit others' text, but it makes sense to include in a policy for now when the status quo is people can just use LLMs and… everyone just feels uncomfortable about it because they don't know where the line for enforcement is.

Reviewer: > FWIW, part of the goal of this comment:
> 
> > To keep things focused on policy, there are two broad categories of comments we'd like to request you avoid:
> > 
> > * Simply stating your viewpoint on LLMs, even if you provide reasons. While these arguments can be useful for the RFC, they are better worded as explicit suggestions to specific areas of the RFC, rather than as just general comments.
> 
> Is mostly to avoid situations where things devolve into a discussion on whether ethics ultimately make usage irredeemable or not, since I don't think that it's going to go particularly well on the RFC. For what it's worth, I agree with you, although reiterating what I mentioned on fedi, the goal is to:
> 
>     1. Push the bar in a direction that focuses on accountability (if you do use an LLM, be honest about it and don't make the code worse for it)
> 
>     2. Put the ethical arguments on the table so that they can be used later to potentially strengthen the policy to reduce LLM usage
> 
> 
> Right now, I'm only aiming for a low bar mostly because I feel like it's most likely to be accepted, and right now there's kind of a lot of open questions that are causing headaches for reviewers. Having _a_ policy in place will make it easier to change the policy, etc.
> 
> Also, FWIW, I do explicitly outline in the policy that usage is not encouraged, only allowed for pragmatic reasons.

Oh, I apologize, my mind completely skipped over that.

I see what you mean, though I fear that once a policy is adopted, there will be much less of a drive to amend it to make it better.

> To respond to a few other points of feedback:
> 
> > I would also explicitly add that the policy covers GenAI images, video, and everything else. The rules for code and other GenAI should not be different. In my views, we should be in solidarity with other affected professions in standing against it.
> 
> So, right now the policy _does_ clarify that both text and code are included, although I didn't mention images, video, and audio since it's not particularly relevant here. However, I do think that this is kind of a weak argument and it's good to be specific.

The reason I feel so strongly about this is that all of these uses of GenAI have the same drawbacks, so it's much more effective to organize against it as a bigger group, rather than only as developers. Though, this is I guess less so about policy, and more so about making a coherent movement against all of GenAI.

Kind of like a nod towards other professions that we are on the same side.

This is more so a nice-to-have for me, rather than a requirement, though. So it's fine to drop it.

> > Accessibility-wise, I am fine with a carve-out, as long as it is not too broad.
> 
> This is kind of what I aimed to go for with the trivial usage definition; the goal is that if the end result was just accessibility, or something that otherwise is indistinguishable, we probably shouldn't be wasting enforcement bandwidth on trying to litigate that. Additionally, it pushes people to understand what they're really using LLMs for, which hopefully will lead to people understanding their effects better.
> 
> Wikipedia's policy is one mentioned in the prior art that stands out pretty well: they allow LLMs for copy-editing but my understanding is that this is mostly pragmatic for enforcement reasons, and they can't really preserve a tone that condones LLM usage when things are publicly editable. Things are different here since you can't just edit others' text, but it makes sense to include in a policy for now when the status quo is people can just use LLMs and… everyone just feels uncomfortable about it because they don't know where the line for enforcement is.

I guess I'll concede with the trivial usage case and not make a problem out of it, as long as the scope of that stays very limited. Banning this might be more effort than it's worth, possibly. Would be nice to discourage its use for non-accessibility cases, here, though.

Non-trivial use is going to be the most harmful, and I would rather completely ban that, rather than just require disclosure.

Reviewer 2: The RFC provides a *lower bound* of what is disallowed on non-trivial use. It requires disclosure, it doesn't say disclosed non-trivial use is always OK.

Reviewer: > The RFC provides a _lower bound_ of what is disallowed on non-trivial use. It requires disclosure, it doesn't say disclosed non-trivial use is always OK.

Yes, and I think it should be expanded to ban all non-trivial use.

Reviewer 3: Perhaps we should ban all computer use entirely. After all, they take away jobs that were historically done by human computers, they use lots of energy, and their manufacture is environmentally damaging.

Actually why stop there, perhaps the problems all began with the printing press?

Boycotting new technologies has never worked, has historically resulted in very poor outcomes for those boycotting it, and in retrospect seems idiotic.

Technology causes change whether you like it or not, and people are harmed when they are not supported in adapting to the change. However, trying to resist the change is futile and causes *more* harm, because it delays adaptation. Effort would be better spent shaping how the technology is used and smoothing out the change, rather than trying to resist it entirely.

Reviewer 4: > Boycotting new technologies has never worked

Demonstrably false; see nuclear energy, asbestos, HFCs, blockchain, NFTs, metaverse, Google Glasses, Stadia, etc. etc.

> Technology causes change whether you like it or not, and people are harmed when they are not supported in adapting to the change. However, trying to resist the change is futile and causes more harm, because it delays adaptation. Effort would be better spent shaping how the technology is used and smoothing out the change, rather than trying to resist it entirely.

A perfectly reasonable take! But I think you're confusing society for its members. This RFC _is_ a part of that conversation around shaping how this technology will be used in the future. History only looks like a straight line in hindsight.

Reviewer: > Perhaps we should ban all computer use entirely. After all, they take away jobs that were historically done by human computers, they use lots of energy, and their manufacture is environmentally damaging.
> 
> Actually why stop there, perhaps the problems all began with the printing press?

There are trade-offs everywhere, there is almost nothing that does not have downsides.

I really would love to see the whole field of computing go in an ethical and sustainable direction, but that is not the situation we live in, right now. I am trying to push the whole industry in a better direction, by doing many things. Resisting GenAI is only one part of it.

> Boycotting new technologies has never worked, has historically resulted in very poor outcomes for those boycotting it, and in retrospect seems idiotic.
>
> Technology causes change whether you like it or not, and people are harmed when they are not supported in adapting to the change. However, trying to resist the change is futile and causes _more_ harm, because it delays adaptation. Effort would be better spent shaping how the technology is used and smoothing out the change, rather than trying to resist it 

I do not subscribe to this defeatist logic. I want to do whatever is practically possible to go against things that are harmful. I'm against proprietary software, I'm against how wasteful the whole tech industry is, I'm vegan, etc.

If something is not okay, we should not just give up.

Reviewer 5: Hi, this is the moderation team of the Rust project: this isn’t the right place to turn to a broad / general discussion about technology; I've collapsed the last 3 replies for this reason and ask you all not to continue that line of discussion.

---

As a general remark on this review thread: This RFC isn't even the right place for having too broad discussions about the topic of “LLM use in Rust”; we should *stay focused to exchanges that do in some way still relate to the RFC at hand*.

We're already having broader discussions on that matter in private channels for rust-lang members, and (even with that limitation on who can participate) it's still so much volume that it's becoming hard to follow; so I unfortunately also can't offer any alternative place where everyone else can participate in broader and/or longer discussions on that topic. *[We're also still investigating ways to make as much of this as possible accessible for others to at least read.]*

Author: Commenting here as I resolve the thread: given that I'm planning to incorporate some of this feedback into the policy and reword it to reduce non-trivial usage, and more strongly condemn LLMs, I think that it's worth just resolving this, especially given the moderator concerns. I was initially keeping it open because the initial feedback was still not incorporated, but now it is.

Author: Since this change is big enough to be worth noting in a separate thread:

in short, non-trivial usage is now disallowed. Disclosure now acts as a way for contributors to justify that LLM usage did not affect the output, i.e., was trivial.

The main motivation for this was the recent news that the Colossus data centre in Memphis, explicitly cited by this policy as an *outlier* because the industry has not been so overtly evil, now powers Claude Code, the model that basically everyone on GitHub is using. Now, it's clear that most LLM users on GitHub are directly doing harm harder and faster than before.

Sure, there are other models, but I doubt many people will switch. The arguments laid out here *do* explicitly support a more restrictive model regardless of this news, and this is only additional justification for going all-in.

It's impossible to look at all the things the AI industry does in a good light. The only possible way is to argue that the ends justify the means, and *even then*, I would say that anyone who isn't a climate denialist and a racist would at least consider switching away from Claude Code toward a different model.

Overall, the reception of this RFC has been overwhelmingly positive, and I'd like to thank the numerous people who've personally reached out to thank me for what I've done so far. You're the folks who make this worth it.

There may still be a few parts that feel weird to read in light of these revisions. As always, please feel free to point them out on the diff.

Reviewer: I believe the summary may need to get updated to add that non-trivial usage is not allowed. Currently it is not very clear to me.

Love how the RFC turned out.

Author: Right, you're totally right. There's also another comment on that I can address too.

Author: Going to also add this in a separate thread because I'm frustrated and don't feel like I have any other outlet to speak about it than this. And I think this is important context for this policy specifically.

The discussion on policy has been absolutely, utterly vitriolic. The people who have been downvoting my comments in Zulip and on GitHub to show dissent instead of just directly insulting me have been really grading. I'm tired of having to be the one to point out the obvious bad faith in the people arguing against reasonable policy and being treated like the villain for it. Honestly, if I were only starting to write a policy now with the amount of hate I've experienced thusfar, I would have just fucking given up; I'm only continuing by momentum and spite at this point.

(Maybe that's not true, but I dunno. I love Rust too much.)

But of course, because said hate is only *implicit*, and not *explicit*, nobody will ultimately receive any consequences from it. I'll just have to deal with it all. I won't even get any fucking apologies.

I can't help but point out how obviously punching down it is for people to pick on the person who's been thrown in the dirt by the AI industry and to claim that really, they're the one who's at fault for trying to limit the harm of LLMs. Really, I'm being insensitive and unreasonable by asking a community that has made me enjoy programming more than I ever have grow a fucking spine and not destroy the thing I love.

Ultimately, this discussion has destroyed our mental health for past month and I know that the same people who are completely unwilling to admit that LLM usage has consequences are similarly unwilling to admit that their actions have taken a toll on us too. This discussion in particular has been a bit of motivation to focus on more healthy plurality, which ironically has been less healthy because the person who wants to just tell everyone how rude they're being is currently not allowed to do so. If you wanna read more, you can check out our blog, but I don't feel like making it easy for the people who've been effectively harassing us to read something so emotionally vulnerable.

I hope you're all happy for what you've done, and I hope you enjoy your time at Rust Week without me. I could have figured out how to stay with a friend in Utrecht, but I genuinely don't think I could look some of you in the face for how foul you've been these past few weeks. And, I'd rather any resources we do manage to get for international travel get spent on trying to reunite with our partner that we haven't seen in 7 years.

You know, something that hasn't been helped by being thrown under the bus by the AI industry and not having a full-time job for 4 years.

Anyway, I've been expecting this to not have any more traction due to Rust Week distractions, but I actually initially messed up my timing and thought that it was this week, not next. So, I get to be even more stressed out about the discussions happening that I won't be participating in. I've wanted to potentially get more involved with the actual team despite only being a meagre helper on the hashbrown repo, but this situation has been just too stressful to work on anything else too seriously.

Have fun, I guess.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Reviewer 6: Whatever it means to you to hear this from the outside: I sympathize greatly with your frustrations here and I appreciate your work more than you can imagine. Please take care, please feel better.

Reviewer: Thank you for all the work you've done. I also really appreciate it.

Take a break if you need to, take care of yourself. If you wish to vent, feel free to reach out.

Wishing you the best.

Author: So, it appears that *somehow* the root of this thread got hidden because of a direct report to GitHub, which is extremely concerning. For now, you can read it on [`rustbot`'s view](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3959#discussion_r3249769422), since it appears that GitHub implemented this (likely new) feature *very* poorly. But note that not even the moderators can unhide this comment, and I'm not even sure org admins can either. Will have to investigate, but, since Rust Week is this week, might take a while.

Reviewer 7: So I understand that my voice does not have very much weight here, since I have only made fairly minor contributions to rust. I do not currently intend to make any future contributions due to the project's external appearance of support of LLMs. I am very strongly in support of this policy, especially now that it has been changed to ban non-trivial LLM use entirely. If it were accepted I would consider contributing to the project again. This is both because of the requirements it would impose but also, more importantly, because it takes the extensive ethical issues seriously rather than pretending that they do not exist or are somehow irrelevant to the practice of software-making. Seeing this gives me hope.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Reviewer: I also have been thinking of contributing for a while, and having a policy like this would be very encouraging for me. I want to contribute to projects that care about ethical issues.

Reviewer 8: As another contributor who hasn't done that much¹, I am in the same boat.

¹ a couple refactors towards const generics during libs blitz, some std API improvements / additions

Reviewer 9: thirded, I find pro LLM stances extremely offputting, have made minor contributions to Rust, rust-analyzer etc. in the past, and am absolutely not motivated to further engage with an ecosystem that is welcoming towards this technology.

Author: Also just clarifying, I explicitly mentioned I appreciated this being posted *despite* it potentially falling under:

* Simply stating your viewpoint on LLMs, even if you provide reasons. While these arguments can be useful for the RFC, they are better worded as explicit suggestions to specific areas of the RFC, rather than as just general comments.

I think that pointing out the sheer number of people who have been outright turned away from this project because of its inability to reject LLMs and their harm is important. So, as long as people keep their comments on this specific point confined to this thread, I think it's fine.

Reviewer 10: I'm in much the same position. I won't contribute until a policy at least as strict as this one is adopted, and honestly even then, I'd still be hesitant to so long as pro-LLM voices are in positions of power in the project.

Reviewer 11: I'd like to voice a similar sentiment: I started learning Rust seven(?)-ish months ago, and I've rapidly fallen in love. while I haven't contributed yet, I became a core contributor to/maintainer of the last language I fell in love with, and have already contemplated contributing to Rust once I have a bit more experience reading and writing it. if this policy or another that is similarly strict against the use of LLMs is adopted, and I otherwise feel safe in the community, there is a good chance I will contribute. if Rust adopts a policy that is permissive of LLMs, I assuredly will not contribute

Reviewer 12: I haven't contributed to the language myself, but Rust is a language that I really like. It's been my go-to language for personal projects for the last 6 years, it's a language I advocate for at my workplace. And seeing https://github.com/rust-lang/rust-forge/commit/14956c3d6c0f8f6ad8706aca2c02c8c67aed3237 made me rather disappointed.

Anyway, just wanted to say thanks to those of you who have contributed and are taking a stand right now. You are appreciated. ❤️

Author: Going to start a separate thread to respond to @ahicks92's comment because I think it's an reasonable comment to be discussed and want at least the folks who can manage with GitHub's UI to keep the thread organised. Link here: https://github.com/rust-lang/rfcs/pull/3959#issuecomment-4462952900

Sorry for the extremely soapboxy ramble. I'm terrible at writing anything that isn't a small essay because I think there's always a lot of context to explain and I tried to strip a lot of it from the RFC *even though* it's so long as-is.

First, and it goes without saying, that the tech industry has done an absolutely abysmal job with accessibility. The fact that GitHub, despite claiming to have various accessibility options, is still horrendous to use even for a sighted person, is just one example of this. (And to anyone, including Austin, who reads this and doesn't want to deal with the threaded UI due to various accessibility issues, please feel free to similarly respond directly in the thread. We can cope.)

I think that Rust itself should do way better to support blind people. One obvious example is that the compiler currently relies a lot on outputting graphical characters to point out spans and such when, well, this is pretty meaningless to a blind person unless you have a very expensive *and big* braille display capable of helping you understand said markings. I've seen loads of projects attempt to create "accessible" CLIs that always word things so they're read properly by screen readers, and try to reduce noise, but ultimately I'm not 100% sure what the best solution is. Right now, I assume that the best way to interface with Rust as a Blind user is just to use rust-analyzer and an IDE that you like, since that will act like a proper UI that you can navigate. Even just being able to go through compiler warnings as a proper list instead of just a bunch of lines on a terminal is an improvement.

One of the explicit accessibility carve-outs I make here is also for stuff like dictation software and text-to-speech, since even though I think it's extremely frustrating that the best possible tools we've designed have been with LLMs, I still think that people should be allowed to use the best possible tools and it's not on them that the best tools are shit.

And a good example of what I mean by this, by the way, is carykh's project from *six years ago* where he managed to make automatic lip-syncing software with some very simple code: https://www.youtube.com/watch?v=y3B8YqeLCpY

While there are some papers that demonstrate that we know so little about glyph recognition that terrible neural networks beat the best hand-built models in performance, this is absolutely *not* true when it comes to audio. We have an astonishingly deep understanding of how human speech works and just need to tap into that linguistic knowledge. The same is true for lexical analysis, and one argument that I left entirely out of this policy because it was off-topic is how current AI models *flagrantly ignore all this research and do the shitty things*. Old-school lexical analysis is able to perfectly tokenise text in basically every language, whereas OpenAI's tokeniser, at least last I checked, *splits Chinese characters between UTF-8 boundaries*. A proper Chinese tokeniser would *at least* try and split up characters into radicals using the data Unicode provides, but this just converts Chinese into byte gibberish that isn't even valid UTF-8.

Basically, it's an absolute condemnation to the tech industry that these are some of the best tools we have available, since the evidence shows that we just aren't trying hard enough. We decided to let hyperscaling and mass surveillance solve the problem instead of our brains.

Anyway, my point here is that I absolutely think Austin and Blind folks have the right to continue using LLMs to make accessibility tools despite the fact that they are an objectively terrible tool for the reasons mentioned in this RFC. It doesn't matter if the best tool for the job is a bad tool; it's a tool, and you can't build without tools. Also, while I think it's a disgrace on the industry that Blind folks are not a large part of it, I don't think that we should consider their financial contribution to the LLM industry as anything worth worrying aboutv.

But, importantly, I think that most of us should not only do better, but actively work to replace these tools with ones that are better *and* more ethical. Like, one thing that constantly rubs me the wrong way is the insistence that LLMs are a tool like any other, when like. My entire point is I very strongly disagree with you. No other tool has managed to so successfully make it difficult to get a job, double the cost of power, make it impossible to buy a computer, and pollute black communities with such efficacy. The fact that we are *literally* building tools and are considering these things as equally useful is a condemnation *of us*, because we are the ones responsible for building tools.

So, my opinion is that while I'm disappointed to see how many things that LLMs are doing for Blind accessibility, this only makes me want to double down on my position that we should continue doing work to reduce LLM usage as much as possible.

The main goal behind the trivial usage provision is twofold, and I discussed this a bit on fedi earlier when summarising the newer policy. The first is hoping that some of the most uncritical LLM users will just let us know who they are so we don't have to ask them first. The second is keeping people as accountable as possible to the code they write to ensure that the LLMs are not writing the code, but they are. Rather than try and guess this, we ask people to tell us explicitly.

I don't know something like a chatbot is the most accessible thing for Blind people to use longer-term, but I certainly don't think that LLMs are the best long-term. The trivial usage clause, to me, is inherently in favour of these kinds of uses, as long as the LLMs aren't writing *non-trivial* amounts of code. Syntax boilerplate, trait implementations, etc., I think are all acceptable usages of this thing, and an old example I got rid of to avoid link rot was the implementation of `&mut I` for `Iterator`: this is sure a lot of text, but it's not exactly intellectually thrilling activity, and I am fine with a machine writing it and checking it for obvious mistakes. What I don't trust a machine to write are all the actually important bits, like real iterator implementations. I do acknowledge this is fundamentally less accessible than allowing an LLM to do it, but I think that we should be aiming to reduce the net harm to everyone done by the AI industry and continue to improve accessibility without it.

Ultimately, that means that some people are going to be unable or unwilling to contribute to Rust, and that's not an insignificant cost. So, we need to pick up our accessibility game and make better tools that don't use LLMs too.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: Also just to comment: I think that the original post's comments on a few topics, like pronouns, were unacceptable, but I think that the issues for Blind Rust users are very real and wanted to address them seriously. Also, since the user said they likely weren't going to comment again, I don't think it's worth trying to litigate those points, since you can't change your behaviour if you don't plan to behave… again? This analogy doesn't work at all.

Just wanted to point this out since I admittedly missed this point on first reading, and think it's worth clarifying that the COC is still relevant during this discussion.

Reviewer 13: > Old-school lexical analysis is able to perfectly tokenise text in basically every language, whereas OpenAI's tokeniser, at least last I checked, splits Chinese characters between UTF-8 boundaries. A proper Chinese tokeniser would at least try and split up characters into radicals using the data Unicode provides, but this just converts Chinese into byte gibberish that isn't even valid UTF-8.

> [..] We decided to let hyperscaling and mass surveillance solve the problem instead of our brains.

This isn't a brainless choice.

First note that the model needs to input and output arbitrary Unicode text, and Unicode has ~300k characters.  There are only 200k tokens to assign; this count is optimized for model performance and hard to change, at least judging by public research.  Therefore some characters must be split into multiple tokens.

This could be done with any encoding scheme.  But for general Unicode, using UTF-8 byte sequences is a natural choice since they’re a reasonably efficient encoding scheme already supported by programming languages.

The question is then whether to special-case Chinese characters to split by radicals instead.  Unicode doesn’t define an actual decomposition for this; there is no way to split a character into a sequence of multiple codepoints that is rendered the same as, or considered equivalent to, the original.  But Unicode does define “ideographic description sequences” which can be used to _represent_ how a series of radicals should be combined into a character, and there is a Unicode-adjacent database mapping Chinese characters to these sequences, so in principle one could define a custom encoding based on that.

I suppose the benefit would be that the model would get an indication of which characters visually resemble which other characters, which could help it understand the language, given that radicals are (loosely) connected to meaning and reading.

However, most Chinese characters in common use already have their own dedicated tokens[^1] and don't need to be split.  One of the main goals of LLM tokenization is to represent text in as few tokens as possible, so if a character doesn't need to be split, it shouldn't be split.  Therefore, radical decomposition would only apply to uncommon characters; therefore, the model would only get an indication of visual resemblance between two characters if both characters are uncommon, which doesn't sound like it would have much benefit.

Perhaps this could be worked around by somehow manipulating the embeddings to reflect decomposition even with single tokens.  I wouldn't be surprised if that happens someday (or even is already happening behind closed doors).  But that wouldn't be an application of existing research.  It would require a lot of new research to determine the best way to do that and evaluate the costs and benefits.

[^1]: According to some random Internet source I found, 98% of written Chinese uses only the top 2500 characters.  From a (separate) list I found of characters sorted by usage, out of the top 2500, 2046 are represented as single tokens in OpenAI's current tokenizer (I measured this using `tiktoken`).

Author: So, I'm going to resolve this thread and the comment associated with it under the guise that I don't think there's really anything to be done. But I edited the comment to point back to this thread.

Author: Procedural update: I updated the summary to properly match the updated policy better, and made some of the small changes I thought were easy to make for now.

Right now, all outstanding requests have been resolved, although so far, most have involved the motivation sections. I would love to see more feedback on the policy specifically, and a clarification of what the folks on the LC would like to see changed before its adoption.

I continue to receive a lot of feedback that this RFC is extremely long, and well, I agree. You're more than welcome to skip on the motivation sections if you're already convinced. But from my understanding, many people *are not convinced* by default, and thus, I have decided to provide *overwhelming evidence* to support the policy.

If major stakeholders in the project, particularly those who do not or did not think that an LLM ban would be well supported, would like to comment on either:

1. if the RFC did convince you to adopt this policy, what was the most convincing motivation?
2. if the RFC did not convince you, why not?

I would appreciate it. Because as it stands, the main reason why I've been hesitant to trim anything here is that a majority of the sections have had at least one person reach out stating that they really felt like a point they wanted to make was addressed *by those sections*.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..97789bca99777b3c370ee46a73a32bea2557c9b7)*

Reviewer 14: I’ve been favoring a restrictive policy for reputational and possible legal reasons, but the part of the motivation that most convinced me was about the Colossus data center showing how the environmental effects of AI disproportionately harm minorities.

Author: Hello, everyone!

You may have noticed that all the active discussion threads have been resolved, minus two:

* One thread for people to express that they would not be comfortable supporting the project if it does not restrict LLM usage.
* One thread where, upon sharing my personal exprerience with the vitriol surrounding this policy discussion, GitHub decided to block my comment under the "acceptable use" policy, even though it does not. They've refused to even respond to my support requests, *even after* it has been privately brought up through other channels to them.

Both of these remain relevant throughout the policy discussion and emphasise the need to put our feet down and unilaterally condemn and restrict LLM usage. Note that I still say *restrict*, not *ban*, because LLM usage is entirely unrestricted *up until the point where it affects the content being written*. If LLMs are still useful for accessibility, or if they are still usefull for sifting through data and other tasks, *these usages are unaffected*. See the RFC itself for more motivation and details.

If you're wondering where the old Motivation section went, a lot of people complained that it should be a blog post, [and now it is](https://txt.ltdk.xyz/sloppery-slope/). Note: effectively all the links that were shared in that motivation section are *still included* as relevant citations, summarised in the newer, shorter motivation section, and the blog post version is linked alongside it for context. I decided to concede that this context was an unnecessary barrier to reading the RFC, *but* importantly, all of the relevant citations are still included. Technically, the justification is all there, just in a form that offers additional reading instead of constituting required reading.

Side note: unfortunately, because GitHub doesn't like accessible semantic HTML, specifically on the GitHub view, you will need to expand the citations explicitly for links within them to work. (In plain HTML, `<details>` are automatically opened when anchors within them are linked. GitHub hates this for some reason.)

The RFC now is almost entirely the actual policy, a timeline of events in the Rust project, and examples of other policies in effect. There are only minor wording changes to the actual policy, plus one additional clause: `Co-Authored-By` trailers are explicitly forbidden, and `Assisted-By` and similar trailers are *discouraged* but not forbidden. This is for two reasons: one, that the presence of these trailers generally indicates nontrivial usage, and two, that they effectively provide advertisements for LLM providers in contributor lists.

Side note: I won't talk too much on the outstanding discussion to create a committee to shepherd the policy discussion. It is my opinion that this policy can continue to solicit feedback from project members regardless of who ultimately ratifies it. I believe that any accepted policy will have a shape similar to this one, similar to the proposed Forge policy, etc. and also believe that the committee should work with existing proposals rather than try to shut them down, which has already been expressed in that discussion.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..5a6f5753e186c228701de343943119f92b7b7acb)*

Author: Messing with the GH API, I managed to find a way to isolate the counts of members who have reacted to the RFC (at the top level) from T-all (with :heart: being used if both :+1: and :heart: are present):

:heart: ×12
:-1: ×3
:+1: ×2

Don't wanna explicitly call people out given the small numbers, but just in case folks are worried that the overwhelmingly positive response is purely due to outside observers: it's not.

<details><summary>For reproducibility</summary>

```shell
curl https://team-api.infra.rust-lang.org/v1/rfcbot.json | jq .all_members[] > members  # new API!
gh api /repos/rust-lang/rfcs/issues/3959/reactions --paginate --jq '.[] | { login: .user.login, content }' > reactions
jq --slurpfile members members --slurpfile reactions reactions --null-input '[$reactions[] | select([.login] | inside($members))] | group_by(.login) | map({key: .[0].login, value: [.[].content]}) | from_entries | map_values(if length == 1 then .[0] elif . == ["+1", "heart"] then "heart" else . end) | to_entries | group_by(.value) | map({ key: .[0].value, value: [.[].key]}) | from_entries'
```

</details>

Author: Another small update: @jyn514's work on creating an `AGENTS.md` file to help enforce the current policy for `rust-lang/rust` demonstrates that models can be pushed to encourage users to disclose and even reduce the extent of their usage, and so, should be allowed within this policy. I've added a dedicated section for this to encourage similar work.

You can see the work on this here: https://github.com/rust-lang/rust/pull/160997

#### Permissive

[LLVM](https://llvm.org/docs/AIToolPolicy.html) requires a "human in the loop" but does not require explicit disclosure. It also explicitly allows a [Bazel Fixer bot](https://discourse.llvm.org/t/rfc-ai-assisted-bazel-fixer-bot/89178/93) which uses AI. They reiterate that contributions can be [extractive](https://llvm.org/docs/AIToolPolicy.html#extractive-contributions) and ask contributors to consider the effort required to review.

[Python](https://devguide.python.org/getting-started/ai-tools/) disallows slop, reiterates requirements of quality, and suggests disclosure but does not require it.

[Firefox](https://firefox-source-docs.mozilla.org/contributing/ai-coding.html) reiterates that humans are responsible for changes but does not require disclosure.

#### In progress

The following projects are currently discussing policy, but have not yet adopted it:

> * [NixOS](https://github.com/NixOS/nixpkgs/issues/410741)

Reviewer 15: Actually this one (NixOS/nixpkgs#514587) has been merged, the current guideline https://github.com/NixOS/nixpkgs/blob/b1f14e1223db7323de9f02986d750b0558f44781/CONTRIBUTING.md#automationai-policy mentioned:

> Every contribution [...] must have a **responsible person in the loop** [...], and must **transparently disclose** any non‐trivial use of automation [...].

My understanding it that it is in the **"Disclosure-required"** category.

* AI generated content (code, discussion, ...) is allowed, contributor is responsible for correctness and legality.
* These generated content must be manually reviewed and/or programmatically verified; vibe-coding without review is banned.
* The AI tool (and verification tool, if any) must be disclosed. For commits they must use the `Assisted-by:` trailer. `Co-authored-by:` is not considered a valid discloure.
* Deliberate violation is considered CoC-breakage.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..b7227e7a3bdd09e800b21d97ee4357f89e8d2d1b)*

Author: Thank you for following up on this too (I'm reviewing in reverse chronological order so this is technically the first comment, but eh)

* [Emacs](https://human-emacs.org)

The following policies exist, but are not final:

> * [Blender](https://devtalk.blender.org/t/ai-contributions-policy-proposal/44202) (disclosure-sometimes-required)

Reviewer 15: The Blender one lives in https://developer.blender.org/docs/handbook/contributing/ai_contributions/. The forum thread said it is the "3rd version". I think this can be thought as "final" even if there may be a v4 in the future. These are always evolving anyway.

Summary:
* Must have human in the loop.
* `Co-authored-by:` banned.

No mention of "discloure" (that part seems taken out in v2). I'd say it is now of **"Permissive"** category.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..b7227e7a3bdd09e800b21d97ee4357f89e8d2d1b)*

Author: Thank you for following up on this; I'll update later

> * [Debian](https://www.debian.org/vote/2026/vote_002) (restrictive)

Reviewer 15: There are now 5 proposals listed there with basically equal number of backers, let's not attribute it as "Restrictive" yet. Discussion started at 7/24 and is expected to last ≥ 2 weeks before voting.

1. **A (restrictive)**: expressly forbid LLMs/AI tools 
2. **B (disclosure-required)**: allowed because taking a neutral stance.
3. **C (partially-restrictive)**: LLM discouraged, default to ban; subprojects may allow LLM due to practicality but disclosure is required
4. **D (disclosure-required)**: allowed because banning is impractical.
5. **E (permissive)**: basically sets no rule aside from having a human responsible; disclosure encouraged but not required.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..b7227e7a3bdd09e800b21d97ee4357f89e8d2d1b)*

Author: Yeah, to be fair I didn't actually read far enough to properly note that there were actually several concurrent proposals

I'll just remove the proposed stance when I get home later since otherwise the point is still that it's not decided yet

1. If the LLM usage is *trivial*, it is completely ignored by the policy and always allowed. Generally, this means that changes made by LLMs are indistinguishable from those made by humans, where the LLM didn't have any creative input into the change.
2. If the LLM usage is *slop*, it is considered spam and moderated accordingly. Generally, this means submitting changes made by LLMs with minimal human intervention.
3. Any *potentially non-trivial* LLM usage must be *disclosed* in ideally as detailed as a manner as possible. This may necessitate additional tooling to notify new contributors about the policy and explain how disclosure works.
4. If a contributor does not fully understand the code they submit, their contribution may be rejected for that reason alone. Note that such usage is not always considered *slop*, and is considered separately. (For example, they may understand a large portion, but not all of it, which shows that they still put in a lot of effort.)
5. If a user is found to be repeatedly lying about LLM usage (using LLMs in a non-trivial way without disclosing that usage), this is a COC violation that will be moderated accordingly.

> 6. As long as users are honest, follow the COC otherwise, and respect boundaries, the worst punishment for *non-trivial* LLM usage is the rejection of a contribution.

Reviewer 7: ```suggestion
6. *Non-trivial* LLM usage is disallowed with a few exceptions, but as long as users are honest, follow the COC otherwise, and respect boundaries, the worst punishment for non-trivial LLM usage is the rejection of a contribution.
```

nit: the original wording I think implies that the policy is looser than it actually is. if the only thing we're saying about non-trivial usage that follows other requirements is that the "worst punishment" is X, it reads to me like there is a wide range of outcomes for typical non-trivial usage. this is weaker than what we're actually doing: banning non-trivial usage by default with a few specific exceptions.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: Right, I actually didn't properly fix this when rewording the summary. Admittedly it's the last bit I edited it and I tried to make the edits minimal, but admittedly too minimal.

Author: Decided to rework the summary entirely, incorporating this change better.

[pcpartpicker.com]: https://pcpartpicker.com

![Average RAM Price (USD) Over Last 18 Months (DDR5-5600 2x32GB) - pcpartpicker.com](https://cdna.pcpartpicker.com/static/forever/images/trends/2026.05.14.usd.ram.ddr5.5600.2x32768.fdf95e58fb5fe98e095cead60afbb5fd.png)

For kits of 2x32GiB DDR5 memory, price went from around 200 USD in November 2024 to nearly 950 USD in May 2026. This is a 5x change in around 18 months.

![Average RAM Price (USD) Over Last 18 Months (DDR4-3200 2x16GB) - pcpartpicker.com](https://cdna.pcpartpicker.com/static/forever/images/trends/2026.05.14.usd.ram.ddr4.3200.2x16384.86e6363cd9987576de00e483293d810d.png)

For kits of 2x16GiB DDR4 memory, price went from around 80 USD in November 2024 to nearly 290 USD in May 2026. This is a 3.5x change in around 18 months.

![Average Solid State Drive Price (USD) Over Last 18 Months (2.5" SATA 1 TB) - pcpartpicker.com](https://cdna.pcpartpicker.com/static/forever/images/trends/2026.05.14.usd.storage.ssd250.1000.fb3f1d55e85d1c9ffff4a3ffd95b947f.png)

For 1TB solid state drives, the price went from around 120 USD in November 2024 to nearly 380 USD in April 2026. This is a 3x change in around 18 months.

> While getting a good computer was already expensive, the AI industry has made it borderline impossible for anyone to get a new computer, or even upgrade an existing one. Even as Rust improves compilation times and memory usage, you can make things work on a slow computer, but you can't make things work on *no* computer.

Author: The "you can't make things work on *no* computer" bit is effectively reinforced by this difficult-to-watch documentary: https://www.youtube.com/watch?v=zyQwAhppWj8

Essentially, the fact that computers have become so expensive means that fewer people are able to afford doing it at all, and it's having ripple effects in the entire industry where people are losing their jobs and companies are going under because nobody is buying a computer.

As I keep seeing examples like this, I find it openly insulting that people see what I've written and summarise it with "ethics cannot affect policy."

This section is one of several that cannot simply be summarised as "ethics" or "politics."

The *least* we can do is to encourage people *not* to use these tools. Don't care if it doesn't help that much. It's the *least* we can do.

(Sharing this here because I'm not sure how I want to incorporate it into the RFC yet, but, it's part of the discussion.)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Reviewer 16: Wow! I guess that makes my desktop's RAM likely worth more than my car, thanks to AI investor craziness! (and I only have 64GiB that I bought a few years ago).

Author: Now linked.

Compare this to supporting, say, cloud service providers. Even if a large amount of money still exchanges hands, in response, we get tools like [docs.rs] and [crater] which are capable of documenting and testing the entire Rust ecosystem. And while you might argue that these tools themselves have their own problems, particularly regarding resource usage, at least we're getting *something useful* out of it.

[docs.rs]: https://docs.rs
[crater]: https://github.com/rust-lang/crater

What do we get out of LLMs that justifies that cost?

### Limits of LLMs

As mentioned before, LLMs are *extremely* sensitive to their source data due to their inability to reason, and this leads to all sorts of issues. One pretty obvious one is that coding models are [pretty bad at operating with less popular languages][esolang], but the even bigger one is that many agentic tools are based upon a fundamentally unbounded feedback loop.

[esolang]: https://arxiv.org/abs/2603.09678v1

> Claude Code, often cited as one of the best tools, has a great example in its source code which it accidentally leaked. Whenever the model is asked to output valid JSON in "non-interactive" mode, i.e. triggered as part of an internal loop, it just repeatedly feeds the model the error in the generated JSON until it correctly generates valid JSON. Originally, the genuine code for this was cited, but it's just way too confusing even with access to the full source code to cite here.

Reviewer 17: This is, to my knowledge, factually inaccurate. While I have not looked at the leaked Claude Code source code, restricting LLMs to valid outputs has been a common feature for several years, [including in Claude](https://platform.claude.com/docs/en/build-with-claude/structured-outputs). llama.cpp has an [open-source implementation](https://github.com/ggml-org/llama.cpp/blob/master/common/json-schema-to-grammar.cpp) of this technique as evidence that this is actually filtering output tokens and not just brute-force. 

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: So, this actually makes a lot of sense and *should* be the way things are done, but at least my best reading of the Claude Code source is it's *not* doing that and I didn't really have any other examples to go off of. And while JSON validation is simple I expect the story to lean in favour of brute force when building code since you can't easily determine whether a particular set of tokens will compile.

Reviewer 18: > but at least my best reading of the Claude Code source is it's not doing that

Masked decoding is implemented in the inference backend as an FSM that triggers on the tool call token. It wouldn't be visible in frontend glue ("harness") like CC. There is some error-looping code in the CC leak, but without the backend code it's hard to tell whether that is necessary or just functionally dead code that a clanker inserted.  I don't think we can draw factual conclusions from the CC source -- with clanker code the thing it "looks like" it's doing is misleading.

Author: Definitely agree that gleaning anything from the source code is probably a worthless endeavour.

Had already trimmed a lot of this section but probably should have trimmed it more.

Author: This section is now gone. I've included the prompts that were previously mentioned in this section in the below one, which now just focuses on social programming within the context of LLMs.

>
> **Corollary (from callstack.md)**: For deadlocks, infinite waits, crashes, and data corruption, "concrete evidence" means proving the code path is structurally possible — not proving it will definitely execute on every run. A `wait_event` with no timeout and no fallback wake condition is a deadlock bug if the wake condition depends on external events that can stop. Do not dismiss such bugs as "unlikely in practice."
>
> This file contains instructions to help you prove a given bug is real.  You must follow every instruction in every section.  Do not skip steps, and you must complete task POSITIVE.1 before completing the false positive check.

[kernel review prompts]: https://github.com/masoncl/review-prompts/blob/main/kernel/false-positive-guide.md

This is, objectively, social programming. And while some of us prefer this way of doing things, many of us just find this way of doing things exhausting. Why does the program need to be told in kind words what *not* to do? How can we be sure that these prompts will always work?

Even if this does work, and it seems to work very well, does that even make it worth it?

> To me, the RFC author, programming is an incredibly fun task, and programming in Rust is especially fun. Writing is fun, too. All of this is a fulfilling creative endeavor more than just a simple means to an end. I simply don't want my life to become this. One of the reasons why I've spent 10+ years learning and writing things in Rust, to the point where I care so much about the community and even write RFCs, is because I think it's genuinely fun. I don't want to outsource that fun to a soulless machine. I don't want to be forced to outsource that fun.

Reviewer 19: While I personally agree with your points here, this paragraph (and the next) reads like a personal blog post, rather than a well-supported position from the Rust org. Could you perhaps reword this so it sounds more like you have gained consensus on the merits of being the human in the loop?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: I'll think about this one; this is pretty much the most personal section of the motivation section, so, I probably should try and make it consistent with everything else. The goal was to be more "I feel this way and know I'm not alone" but, that tone is not the same as the rest of the document.

Author: This part has been mostly removed.

6. As long as users are honest, follow the COC otherwise, and respect boundaries, the worst punishment for *non-trivial* LLM usage is the rejection of a contribution.

### Trivial Usage is Always Allowed

The policy explicitly allows all *trivial* LLM usage, which can generally be thought as LLM usage where the LLM had no creative input in decisionmaking. Colloquially, when deciding whether to label something as LLM usage, all trivial usage is excluded from these decisions, although you're always free to discuss your process for doing things. The following examples constitute trivial LLM usage:

* Usage of speech-to-text models to transcribe speech into text, as long as such transcription is reasonably accurate. In this case, the model was not deciding what to write; the person speaking was.
* Basic auto-completion of syntax, spell-checking, and copy-editing. In this case, the model is simply accelerating what a user already intended to do, rather than deciding what to do.
* Even the writing of certain code or text can be considered trivial, if little creative input is required to write it. "Boilerplate" code is a good example of this.

> Note that this LLM usage being allowed *does not* constitute an explicit endorsement; it simply represents a pragmatic approach to enforcement, since it is difficult to distinguish. This policy does not try to distinguish between generative AI, LLMs, and other forms of machine learning, since the category of "trivial usage" covers broadly enough to avoid needing that distinction.

Reviewer 7: A large fraction of the ethical issues with LLM use discussed above apply equally to trivial and nontrivial usage. It is perhaps worth strengthening "not explicitly endorsing" trivial usage to explicitly condemning it, while admitting that enforcing a ban on trivial usage is less practical than a ban on nontrivial usage and so we may focus on nontrivial usage out of pragmatism.

Personally I think it may be a good idea to go farther and require disclosure for both trivial and nontrivial usage. This makes things simpler for contributors, who must always disclose LLM use rather than being asked to distinguish between trivial and possibly-non-trivial use privately. It's much more difficult to enforce disclosure for trivial use, and we would likely want to err on the side of caution when identifying undisclosed trivial use. However, I believe this is okay. It is okay to ask people for things without an effective mechanism to force them to comply. This is the basis of ideas like trust and consent, and if we are abandoning trust and consent we have bigger problems.

I do completely understand if you don't want to strengthen the requirements around trivial usage because it would make the policy less likely to be accepted.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: So, I think this is interesting to interpret in light of this comment: https://github.com/rust-lang/rfcs/pull/3959#issuecomment-4462952900

and my response: https://github.com/rust-lang/rfcs/pull/3959#discussion_r3251413961

since, ultimately, even though I think it's an absolute condemnation of the industry for not making anything better, these *do* constitute accessibility tools that people use, and we want to make sure that we're specifically condemning the industry and not endorsing the usage, rather than condemning the users.

Also, in private I've made the analogy that we're asking a large industry to not donate half of their salaries to the evilest part of that industry, and not just chastising individuals for their usage of these things for one reason or another. I personally think that the ethical issues make the things revolting to use, but that ultimately doesn't change that they have helped people in one way or another. Rather than just telling everyone that they're being wrong for using them, I think we should try and get them to join our side and fight against the industry, often by replacing the terrible tool of LLMs with a better one, instead of just chastising them.

Reviewer 7: That's a fair argument that explicitly condemning use (as opposed to just disallowing it) is not useful in this context. I think I am now in agreement.

LLM/"AI" use in the context of accessibility is genuinely really complicated and yeah, I don't really have any interest in condemning or disallowing accessibility use-cases. Where you draw the line is complicated though. It would be very bad for the policy to ban machine transcription use in the development process, but I do think we should disallow the kinds of examples @ahicks92 described in their comment where LLMs were used to generate accessibility software rather than as an accessibility tool itself. A nontrivially LLM-generated PR to add screen-reader-friendly error reporting to rust should be rejected just the same as any other nontrivially LLM-generated PR.

I think a disclosure requirement for trivial accessibility use would be harmful because it forces disabled contributors to mark themselves as disabled. Trivial use for non-accessibility reasons (boilerplate generation or autocomplete) is a lot less complicated and I am straightforwardly ethically opposed to it for mostly the same reasons I am straightforwardly ethically opposed to nontrivial use. Ideally I would like to have policy would ban trivial non-accessibility use, or at least require disclosure for trivial non-accessibility use. However, I can see the problem that trying to draw a box around trivial accessibility use at the policy level is much more fraught than drawing a box around trivial use in general. A common kind of harm done by accessibility policies that draw a tight box around accessibility vs non-accessibility use-cases is that they require disabled people to "prove" that they are sufficiently disabled to be able to use a thing for accessibility purposes to a system that is both structurally and individually ableist. We should be very careful to avoid producing this kind of outcome here. I am not sure where I stand on balancing these issues, and when in doubt would prefer to err on the side of not making an ableist system.

Author: Right, and to be clear, I'm still saying that we should disallow ultimately letting the LLM write any code that isn't just shortcutting syntax or something to that effect. Ultimately, the idea here is that if we really taper what kinds of things people are allowed to use LLMs for, the people who are just using them to shortcut actually writing code will either give up or acknowledge that the tool isn't helping them actually write code, just doing it for them, and the people who still continue to use it are the people who feel that it is actually helping for accessibility concerns. Kinda hoping that the current formulation is self-correcting for that, but as usual, no idea what would happen in practice.

Like, I mention trait implementations as a common example in Rust because they can be genuinely lengthy but pretty much have exactly one way to be written in a lot of cases. Personally, I think that we should just get `rust-analyzer` to the point where it's able to autocomplete these trivial cases for us, and then an LLM is not involved, but in cases where it's not possible, I'll relent and just let people do it.

Like, working with code is *atrocious* for a blind person even with the best tools, so, I'm more than willing to give people the benefit of the doubt that these things actually help them make it tolerable.

Author: FWIW: keeping this open to mostly address the "not endorse" question and how that should be worded, as a reminder to look into that later.

Author: So, I'm going to resolve this for now, under the premise that I think the larger question of whether the individuals should be allowed to endorse things on behalf of the project, should be a separate discussion entirely. This kind of mixes in with my other thoughts on encouraging [more comments like this](https://blog.rust-lang.org/2022/02/24/Rust-1.59.0/):

> Today's release falls on the day in which the world's attention is captured by the sudden invasion of Ukraine by Putin's forces. Before going into the details of the new Rust release, we'd like to state that we stand in solidarity with the people of Ukraine and express our support for all people affected by this conflict.

We need to come up with a path which both encourages these types of comments being made front and centre for the project, while mitigating the potential issues that come from doing the same with LLM usage.

They're uncomfortable problems.

But I'm not asking you to stop. I'm asking you to be more careful with how you use it, and to limit the cases where there could potentially be issues. I do feel that, in a very real sense, LLM users are directly harmed by the industry that makes the tool they love. Genuinely, every bit of anger toward someone who uses an LLM is better pointed toward the people in charge doing actual harm. But you must understand why it's hard not to be a bit angry at you, too, even if we try to be nicer when we can.

Hopefully, this policy's adoption encourages us to be a little nicer to each other, and to understand, more than anything else.

### Avoiding the vibe of vibecoding

This one actually feels reasonable at first glance, but unfortunately, it falls short.

Essentially, the idea is that instead of limiting nontrivial usage in shared spaces and public communications, we should simply relax the rule and ensure that users don't replicate the "LLM feeling" in their contributions.

The problem with this is that LLMs are designed to replicate humans, and there have already been numerous issues that have cropped up:

> * Neurodivergent authors tend to replicate the "terseness" of many LLMs, and often show up as false positives in LLM detection

Reviewer 20: Hi, I think this should be reworded. I have never known LLMs to be terse unless explicitly commanded to be so and if you mean something in the realm of bluntness~brusqueness, you should just say that instead of euphemizing it. Using a word to mean something *other* than what it actually means (with the true meaning supposed to be inferred from context) is ironically like the worst possible case for a lot of forms of neurodivergence. If it's not meant to signify that then it kind of proves my point that it's confusing and hard to interpret because the above is my best guess.

I'm happy to provide suggestions instead of just complaining but I'd like to know for sure what the intended meaning is first so I'm not wasting time. Painting with too broad a brush about ND people is obviously bad and to me that's what this wording seems to be trying to avoid, but I think there's likely a way of wording it that's better all around.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Author: Yeah, I'd be more than happy to reword this; I'm ND myself so I get the desire to be careful due to all the horrible ways people tend to interpret stuff.

Might just end up wording it as "replicating some of the quirks" rather than "replicating the terseness" so it's just, totally vague for now. There are even academic studies that have proven this point but I hadn't felt like linking them here since the main purpose of this section is "hey, LLM detection has false positives both for automatic methods and manual methods, and ND people tend to suffer as a result"

This comment also reminds me that I probably don't fully understand what additional connotations come behind the word "terse" since I just thought it was a erm, terser, way to say "to the point"

Reviewer 21: Like everything in English it's contextual, but "terse" can be mildly pejorative. It has the nuance of prioritising concision over other considerations, even if the listener (or reader) wanted more detail. I think it would be ok to leave as is rather than polishing every word in a long essay. but if you want something that sounds neutral-to-positive, an option would be unquoted `brevity`. Alternatively `"directness"` might convey what you were looking for?

Author: For now, I think I'm going to change this to:

> Neurodivergent authors tend to replicate some of the quirks of LLMs, and often show up as false positives in LLM detection

To avoid any undue value judgements, since I do think there are multiple qualities that could be discussed here.

- Feature Name: N/A
- Start Date: 2026-05-04
- RFC PR: [rust-lang/rfcs#3959](https://github.com/rust-lang/rfcs/pull/3959)
- Rust Issue: N/A

## Summary
[summary]: #summary

This RFC proposes a strict policy regarding generative Artificial Intelligence (AI) models, specifically Large Language Models (LLMs), and their use within the rust-lang organization.

It proposes an "honesty over purity" policy where maintainers are given broad authority to decide what amount of LLM-generated code is acceptable, while avoiding repercussions for those who do use LLMs and are honest about it. This can be summarized in the following checklist with terms that will be defined throughout the RFC:

> 1. If the LLM usage is *trivial*, it is completely ignored by the policy and always allowed. Generally, this means that changes made by LLMs are indistinguishable from those made by humans, where the LLM didn't have any creative input into the change.

Reviewer 22: Should this be "by LLMs `that` are indistinguishable"?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..f10599f9b451f5eae846100914ae744d8055cf50)*

Author: Probably not:

Generally, this means that (changes made by LLMs are (indistinguishable from those made by humans, where (the LLM didn't have any creative input into the change))).

2. If the LLM usage is *slop*, it is considered spam and moderated accordingly. Generally, this means submitting changes made by LLMs with minimal human intervention.

> 3. *Nontrivial* LLM usage must be *disclosed* in ideally as detailed as a manner as possible. This may necessitate additional tooling to notify new contributors about the policy and explain how disclosure works.

Reviewer 4: In PRs I've submitted to `rust` and `cargo`, I include a single sentence at the end usually of the form:

> No AI tooling of any kind was used during the creation of this PR.

I think including an AI disclosure prompt/template in the GitHub PR template is a simple way to cover off disclosure while allowing the author the freedom to go into as much or as little detail as they feel is appropriate.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..f10599f9b451f5eae846100914ae744d8055cf50)*

Author: Yeah, having a section on a PR template just to fill in LLM disclosure is a good idea, and that's one of the multiple options that could be used for tooling.

This is actually the one difference between the summary for the RFC and the one proposed for inclusion in the actual policy: for the RFC, the tooling is an important point since it doesn't exist yet, but for the final policy, the goal is to already have that.

I mostly leave the details out of the RFC since they're almost certainly going to be done via experimentation, but this is a pretty easy one to start with.

Ultimately, LLM support has an *unprecedented* ability to fund the AI industry, and the industry is using this power to largely enforce systemic racism by suffocating Black people in Memphis, wringing Kenyan data workers dry, and preventing all but the most Industry-endorsed not-minorities from obtaining a job.

Compare this to supporting, say, cloud service providers. Even if a large amount of money still exchanges hands, in response, we get tools like [docs.rs] and [crater] which are capable of documenting and testing the entire Rust ecosystem. And while you might argue that these tools themselves have their own problems, particularly regarding resource usage, at least we're getting *something useful* out of it.

[docs.rs]: https://docs.rs
[crater]: https://github.com/rust-lang/crater

What do we get out of LLMs that justifies that cost?

> ### Effectiveness of LLMs

Reviewer 4: I personally don't believe the RFC should consider the capability of LLMs in its policies. I agree with your summary, but I believe the ethical and ethics-adjacent concerns are sufficient to justify the policy proposed. I don't believe the policy would or should change if, hypothetically, Claude Code used zero power and made no mistakes.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..f10599f9b451f5eae846100914ae744d8055cf50)*

Author: So, I have a feeling you might be doing the thing I did when reviewing the other policy RFC, which is comment as I'm reading without necessarily picking up the full context. You also could have not done that, but I figured I'd preface this with that since, it might not 100% apply.

This section is technically a bit misleading since it's not *really* about the effectiveness of LLMs, but rather on how their design fundamentally limits how they can operate. I tried in the ethics section to clarify that some of the ethical concerns are things that fundamentally can't really change based upon the design of LLMs (for example, many of them don't really work with less data), but this section also kind of goes onto that by pointing out that due to the fundamental reliance on brute force, they will always have some pitfalls that are hard to avoid.

Like, one big argument on the ethical side isn't necessarily to ignore those problems, but that the ethical problems can slowly be trimmed away on newer and newer models, and I kind of want to push back on that by pointing out that some of these issues are fundamental. For example, a lot of people are hoping to shift to high-quality models trained entirely on permissive code that run locally on your machine, which do already exist today. The point of this section is to point out how even those models will still run into a few fundamental limits. Even if we could get the data without stealing and avoid pollution by running entirely on solar, for example, we'd end up with models that *probably* still can't spell strawberry.

Also, on the subject of covering bases with arguments, I do want to point out this bit from the PR description:

> I would appreciate that these arguments be directed on potentially simplifying the text of the policy itself, rather than just removing sections of motivation.

I mostly reply here because I don't necessarily think this comment is about removing the section, rather, potentially rewording some of it to be a bit more clear.

Reviewer 4: > Like, one big argument on the ethical side isn't necessarily to ignore those problems, but that the ethical problems can slowly be trimmed away on newer and newer models, and I kind of want to push back on that by pointing out that some of these issues are fundamental. For example, a lot of people are hoping to shift to high-quality models trained entirely on permissive code that run locally on your machine, which do already exist today. The point of this section is to point out how even those models will still run into a few fundamental limits. Even if we could get the data without stealing and avoid pollution by running entirely on solar, for example, we'd end up with models that probably still can't spell strawberry.

I may not be articulating it well (I know you and the larger Rust community have been discussing this issue at length and likely have covered off most of these debates already). My concern is that if we tie the policy to the performance of the AI system, it will remain in a constant state of relitigation as the models improve. But as you pointed out the same can be said for the ethical section, so that might be why I'm struggling to make a strong case for one over the other.

After sitting on this over breakfast I think it's fair to mark this as resolved, as I don't think my concerns around keeping the policy effective in a hypothetical "the AI is perfect" world would be addressed at this point in the RFC anyway.

P.S. I did read through the full rendered document first, but reopened it for review and at that point left comments as I went. I choose to blame the late night reading for me missing that note around section removal, sorry!

Hopefully, this policy's adoption encourages us to be a little nicer to each other, and to understand, more than anything else.

### Avoiding the vibe of vibecoding

This one actually feels reasonable at first glance, but unfortunately, it falls short.

Essentially, the idea is that instead of limiting nontrivial usage in shared spaces and public communications, we should simply relax the rule and ensure that users don't replicate the "LLM feeling" in their contributions.

The problem with this is that LLMs are designed to replicate humans, and there have already been numerous issues that have cropped up:

* Neurodivergent authors tend to replicate the "terseness" of many LLMs, and often show up as false positives in LLM detection

> * Kenyan authors, many of whom helped filter the data for LLMs, often show up as false positives in LLM detection

Author: FWIW, while I've tried to cite everything across the RFC, this section was added after the first draft and I knew both of these had citations, but had a lot of trouble finding explicit articles about it. So, if you happen to find sources for these, I'd be happy to update the RFC to include them.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..f10599f9b451f5eae846100914ae744d8055cf50)*

Reviewer 23: When reading that section, I thought back to this article: https://marcusolang.substack.com/p/im-kenyan-i-dont-write-like-chatgpt.

Author: (To be clear, I do think this blog post is sufficient evidence for the second point, although I'm mostly leaving this open to remind myself to look for other sources too.)

- Feature Name: N/A
- Start Date: 2026-05-04
- RFC PR: [rust-lang/rfcs#3959](https://github.com/rust-lang/rfcs/pull/3959)
- Rust Issue: N/A

## Summary
[summary]: #summary

This RFC proposes a strict policy regarding generative Artificial Intelligence (AI) models, specifically Large Language Models (LLMs), and their use within the rust-lang organization.

It proposes an "honesty over purity" policy where maintainers are given broad authority to decide what amount of LLM-generated code is acceptable, while avoiding repercussions for those who do use LLMs and are honest about it. This can be summarized in the following checklist with terms that will be defined throughout the RFC:

1. If the LLM usage is *trivial*, it is completely ignored by the policy and always allowed. Generally, this means that changes made by LLMs are indistinguishable from those made by humans, where the LLM didn't have any creative input into the change.
2. If the LLM usage is *slop*, it is considered spam and moderated accordingly. Generally, this means submitting changes made by LLMs with minimal human intervention.

> 3. *Nontrivial* LLM usage must be *disclosed* in ideally as detailed as a manner as possible. This may necessitate additional tooling to notify new contributors about the policy and explain how disclosure works.

Reviewer: In my opinion, it's best to ban it, rather than require disclosure. A bad thing is still a bad thing, regardless of whether you disclose it or not.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: FWIW, I am considering how this could be strengthened, although I do want to try and find a way for people to be honest regardless. Because under full bans there are going to be people who just decide to then be dishonest and we don't want them clogging up the moderation queue.

This is also the benefit of the trivial usage cause, because it pushes people to say "I used an AI *but* I had full creative control" rather than simply not disclosing and requiring extra back and forth. Although there are genuinely points to consider on whether all this matters because unfortunately everything is just guesswork right now, since we have no idea how people will act after a policy is in place.

Reviewer: People may lie about it, just like they also could be writing the contribution on company time and using company equipment, without the company stating the copyright is given to the author. This case is way harder to detect, but we still cannot accept it.

I believe people lying about it should not be a reason not to enact a ban on non-trivial use. To me, the project should try to uphold the highest ethical standards practically possible.

GenAI code may slip through, and that is of course not ideal, but at least the project is striving to be ethical.

Reviewer 3: I don't believe ethics should come into this at all. What's next, you ban people from contributing who eat meat?

The code of conduct exists because it is important to be inclusive and maintain good interpersonal relationships for the people working on Rust, in order for the project to be successful. It is explicitly not an attempt to push your own agenda on others.

Reviewer: > I don't believe ethics should come into this at all. What's next, you ban people from contributing who eat meat?
> 
> The code of conduct exists because it is important to be inclusive and maintain good interpersonal relationships for the people working on Rust, in order for the project to be successful. It is explicitly not an attempt to push your own agenda on others.

This policy has nothing to do with judging what people do in their personal lives. It's not preventing you from using GenAI in your personal projects. Nor is it restricting you in what you can do in your own forks of Rust projects. It is also not my goal to dictate what you do in your personal life, as I do not think any policy should be doing that.

It's also not my goal to be doing value judgements on people, and I do not think a policy should be doing that, either. Just outline what is acceptable for the project and its users. I am not interested in witch hunts and such, either.

If you do not care about the ethical issues with GenAI, or you disagree that there are ethical issues, that is also your choice and I am not here to tell you what to believe.

(If this is veering too much into an off-topic direction, feel free to tell me to hide this comment, and I will hide it.)

Reviewer 5: > I believe people lying about it should not be a reason not to enact a ban on non-trivial use.

> *[…]* What's next, you ban people from contributing who eat meat?

I’d like everyone to pay attention to the different uses of “ban”. In prior discussions I’ve already noticed that it’s an easy thing to gloss over when using this verb. @clarfonthey did a good job avoiding ambiguities IMO (and there’s not a single instance of “ban” in the RFC Summary section)… but in this comment thread it’s coming up again and (I can’t read minds, but) this might have lead to some degree of miscommunication.

The commentary from @lumi-me-not could/should perhaps more closely relate to the already included point no. 6 in the RFC text, stating:

> 6. Teams are allowed to form their own policies regarding nontrivial LLM usage, although as long as users are honest, follow the COC otherwise, and respect boundaries, the worst that will happen is the rejection of a contribution.

If the idea to be discussed is to raise the baseline from the minimum – possibly up to the maximum – and perhaps even not to allow teams some (or any) freedom in allowing less restrictions – within this frame of possibilities, a “enact a ban on non-trivial use” would still just mean “documenting the practice that certain kinds of PRs will generally be rejected”. It wouldn’t have to mean that any such person will be “banned” in a moderation / GitHub-[blocking](https://docs.github.com/en/communities/maintaining-your-safety-on-github/blocking-a-user-from-your-organization) kind of sense.

[Moreover “rejected” doesn’t even ***have*** to mean that *all* problematic PRs are necessarily being closed; *small* but non-*trivial* use could still be reduced due to review/feedback.]

---

Of course, that distinction of “blocking people” or “rejecting contributions” is completely impossible to make in the “people who eat meat” analogy, which is why I think it’s a rather bad analogy.

Reviewer 3: > If you do not care about the ethical issues with GenAI, or you disagree that there are ethical issues, that is also your choice and I am not here to tell you what to believe.

While you say you are not trying to tell individuals what to believe, the proposed policy still asks the project itself to adopt a particular (subjective) ethical stance and evaluate contributions through that lens, rather than primarily on practical or technical grounds.

This inherently alienates anyone who doesn't share the same beliefs, even when those beliefs are perfectly compatible with the success and harmony of the project.

I think this reflects a deeper disagreement about the role of the project itself: whether it should function primarily as a technical collaboration with minimal ideological filtering, or as a community that uses its position and influence as a successful project to steer contributors toward particular ethical positions through its contribution policies.

Reviewer 24: This statement is a contradiction:

> This inherently alienates anyone who doesn't share the same beliefs, even when those beliefs are perfectly compatible with the success and harmony of the project.

If people who don't share beliefs feel alienated then there must already exist sufficient incompatibility between them to disrupt the harmony.

Regardless, this comment doesn't refute the preceding arguments. Policy doesn't dictate belief because:

1. It isn't necessary to believe in a policy in order to follow it.
2. Policy applies within a defined domain and need not be followed elsewhere.

The position of technical collaboration with minimal ideological filtering goes both ways: one can also contribute objectively in accord with a policy they subjectively disagree with.

Disagreement about the role of the project may go even deeper:  the role and practicality of objectivity in technical fields are by no means settled matters. Objectivity [is itself a value](https://plato.stanford.edu/entries/scientific-objectivity/#Intr) and the [value-free ideal](https://plato.stanford.edu/entries/scientific-objectivity/#ObjeAbseNormCommValuFreeIdea) is by no means universally accepted as [desirable](https://plato.stanford.edu/entries/scientific-objectivity/#SciePoliValuFreeIdea) or [possible](https://plato.stanford.edu/entries/scientific-objectivity/#FeyeTyraRatiMeth). It may be that these disagreements follow from those. But this is straying from the topic.

Reviewer: > I’d like everyone to pay attention to the different uses of “ban”. In prior discussions I’ve already noticed that it’s an easy thing to gloss over when using this verb. @clarfonthey did a good job avoiding ambiguities IMO (and there’s not a single instance of “ban” in the RFC Summary section)… but in this comment thread it’s coming up again and (I can’t read minds, but) this might have lead to some degree of miscommunication.

Yeah, with "ban" I do not mean any moderation actions or punishment. I don't think it's useful or desirable to punish for usage of GenAI. Of course, someone repeatedly ignoring the policy and acting in bad faith could get some punishment, but in that case, GenAI use is not quite what they're being punished for. Rather, it is their conduct.

> The commentary from @lumi-me-not could/should perhaps more closely relate to the already included point no. 6 in the RFC text, stating:
> 
> > 6. Teams are allowed to form their own policies regarding nontrivial LLM usage, although as long as users are honest, follow the COC otherwise, and respect boundaries, the worst that will happen is the rejection of a contribution.
> 
> If the idea to be discussed is to raise the baseline from the minimum – possibly up to the maximum – and perhaps even not to allow teams some (or any) freedom in allowing less restrictions – within this frame of possibilities, a “enact a ban on non-trivial use” would still just mean “documenting the practice that certain kinds of PRs will generally be rejected”. It wouldn’t have to mean that any such person will be “banned” in a moderation / GitHub-[blocking](https://docs.github.com/en/communities/maintaining-your-safety-on-github/blocking-a-user-from-your-organization) kind of sense.
>
> [Moreover “rejected” doesn’t even _**have**_ to mean that _all_ problematic PRs are necessarily being closed; _small_ but non-_trivial_ use could still be reduced due to review/feedback.]

Moving the baseline to the maximum is pretty much what I mean here. That is to say, not allowing PRs that have any non-trivial use of GenAI.

Of course, it might not always be easy to detect, so we should assume people are following the policy unless there is strong suspicion or proof otherwise.

Author: @Diggsey, I find it quite frustrating that you've not only ignored comments from a moderator twice, but that you've also demonstrably refused to acknowledge the content of the RFC and its various points. I don't know how much more direct I can be:

> The RFC *does not* take the position that any of these concerns should unilaterally condemn LLM use. However, because these concerns inform many people's opinions of LLMs, they are worth explaining.

Also:

> Boycotting new technologies has never worked, has historically resulted in very poor outcomes for those boycotting it, and in retrospect seems idiotic.

I'm going to start by stating that I'm *already* in a sour mood today after having to argue with people how [the news](https://www.cnbc.com/2026/05/06/anthropic-spacex-data-center-capacity.html) implicates Anthropic in systemic racism and we should at least *pretend* to care about that. I don't know how you could possibly see concrete evidence that Anthropic is okay helping Grok suffocate a historically black community in the U.S. and reply with anything other than "that's disgusting," certainly not "ethics shouldn't be a part of deciding policy."

You know what's *one* counterexample to your argument? The civil rights movement in the U.S., which directly countered systemic racism. Even if you're not American, there are *countless* movements across the globe which have included successful boycotts, and you must have heard of at least *one* of them. Or, you refuse to acknowledge them. Can't tell which one is more favourable for you: actually ignorant or intentionally so.

Also, Charles Boycott was so successfully ostracised for being such a horrible person that he literally got the term named after him. So no, I'm not accepting your claim that boycotting has "historically resulted in very poor outcomes for those boycotting." Citation needed.

I would be more than happy to respond to any *good-faith* concerns from you regarding the policy or its arguments. However, I haven't gotten the impression that you have any given your latest comments. If you're not willing to read the arguments I made, I see no reason why I should read yours.

Despite having the permission to hide your comments as a member of the team, I've been delegating that role to the moderator team since I'm inherently biased as the author of this RFC. Even though the mods are *literally* asleep right now, I'm still going to honour that for now.

Please don't make me regret my decision.

Reviewer 3: > I find it quite frustrating that you've not only ignored comments from a moderator twice

That was certainly not my intention - I saw @steffahn 's message in the other thread and stopped replying there as a result. However, in this thread @lumi-me-not is suggesting changes to this RFC which would have sweeping repercussions for the project. If my comments are off topic, then so is this whole thread.

It's clear my input is not wanted here, so I will cease commenting, but I feel like I must correct the strawman you are attacking:

>  I'm not accepting your claim that boycotting has "historically resulted in very poor outcomes for those boycotting."

I did not make this claim. I said boycotting *new technologies* has never worked. Absolutely you can boycott events and companies successfully: to take the example mentioned previously, Google Glass may be long dead but so-called "smart glasses" fill a shelf of every shop selling sunglasses.

Author: > If my comments are off topic, then so is this whole thread.

This is extremely incorrect. You completely misrepresent the situation by making useless claims to banning meat when again, *the goal is not to ban LLM usage*, but to reduce its harm. In this analogy, we're asking people to not grossly negligently put their children on all-meat diets, or maybe not eat meat every single day.

I think that the most effective way to reduce the harm from LLMs would be to ban it entirely, but I'm not the dictator of Rust and can't just decide what we do, so instead, I've proposed this comparably lax policy instead.

Your comments are only off-topic in the sense that they show that you have flagrant disregard for the contents of the RFC itself. Again, I think it's only fair that, if you wish to argue with the policy, you should engage its arguments directly, rather than ignoring them and misrepresenting them with straw-man replacements.

> I did not make this claim. I said boycotting *new technologies* has never worked.

I don't appreciate you squirming out of an argument by playing games with words. Sure, if you limit the scope of your argument to something that *technically* works, then you will be *technically* correct. For example, you could argue that the new "technology" involved during the industrial revolution, was not stopped by boycotts. However, the massive labour rights violations they caused were successfully improved by boycotts and strikes. We have weekends because of this.

Similarly, your argument on smart glasses feels in extremely poor faith because you can't announce victory when the fight is obviously not over. Again, if you argue that because they're no longer a new technology, current boycotts don't apply, you're choosing to selectively apply an argument just to win.

This is not even a boycott of LLMs. This is a request to be more thoughtful with their usage and to avoid their negative effects. In this thread, we're debating the benefits of eliminating *non-trivial* usage of LLMs, which has been explicitly delineated in the RFC. You're just calling this a ban of LLM usage, and arguing against "pushing an agenda," which is just completely irrelevant.

Rust *is* pushing an agenda, actually. We argue for memory safety and flexibility and so many other things. We also have a code of conduct that explicitly goes beyond what is reasonable in many ways to be inclusive of other people. As I explicitly argue in the RFC, I think that LLM support is *not* inclusive of other people for a number of reasons. But you're choosing to ignore all those arguments just so you can complain.

Again, I will repeat myself: I would be more than happy to respond to any *good-faith* concerns from you regarding the policy or its arguments. However, I haven't gotten the impression that you have any given your latest comments. If you're not willing to read the arguments I made, I see no reason why I should read yours.

Author: So, marking this as resolved because I *am* working on updating the policy so that this is the case. We can discuss more once I post the changes.

4. If a contributor does not fully understand the code they submit, their contribution may be rejected for that reason alone. Note that such usage is not always considered *slop*, and is considered separately. (For example, they may understand a large portion, but not all of it, which shows that they still put in a lot of effort.)

> 5. If a user is found to be repeatedly lying about LLM usage (using LLMs without disclosing that usage), this is a COC violation that will be moderated accordingly.

Reviewer 25: This all looks great!   "Honesty over purity" is a more friendly way of saying that non-disclosed nontrivial LLM uses is bad faith.

I'd suggest rewording 5 so that a single offence can trigger moderation.  You could just remove the word "repeatedly", but then the sentence maybe stronger than you'd like, since some lack of clarity exists on the first offence.  "repeatedly or dangerously" maybe?  Or some rephrasing? 

Anyway I do think you want to leave yourself room to act strongly on a first offence.  You maybe confident 6 and 7 cover those cases, but I think you're not doing the vibers any favours by giving the impression of a free pass here.  

You should probably change "lying" to "hiding" too, because someone might think that saying nothing is not lying.  There can be explicit questions in PR templates etc that make "hiding" more clearly be "lying", but this policy seems more general, so "hiding" seems clearer here.

Author: So, not 100% sure if you've read the whole thing, since I do go a bit more in depth in the start of the motivation to point out what "honesty over purity" means; nontrivial use is allowed (less purity) with disclosure (honesty). The point is that this is a compromise over a more strict policy where honesty is seen more important than preventing all (nontrivial) LLM usage.

With regard to enforcement, 5 already leaves enforcement largely up to the moderation team, and I would rather that they have broad authority to determine whether something is intentional or not. The main method of determining usage where things might be ambiguous is by asking people directly, and then you're kind of forced to say something about it; at that point, if you lie, that's just lying, not hiding.

Right now there are already a lot of moderation details I kept out of the policy for simplicity: for example, the moderation team tends to prefer bans for slop PRs because bans can be rescinded; the only change is they have to talk to the mods about it before future contributions. For stuff like what to do when lying, I'd rather the mods who are already pretty lenient (intentionally) make the determination whether someone was being malicious or not.

Plus, COC enforcement also is inherently asymmetrical because it's more heavily weighted against team members (who should know better), so we kind of get the desired enforcement built in: team members who lie are treated much more seriously and might not get a second chance, whereas for ordinary contributors who do it, we're more likely to just reject PRs unless it becomes a problem.

Reviewer 25: Alright, these two concerns jumped out as "weak" to me, but so long as everyone feels the overall document gives them enough leeway in moderation then I guess my remark are resolved.  :)

The main issue is forming a policy with regard to all other forms of LLM usage. A large portion of the team have serious concerns regarding LLM usage, but there are also several team members who feel they would be excluded by a complete ban on LLM usage. There's also a pragmatic issue with enforcing any limit on LLM usage, where some LLM usage is simply impossible to detect and is effectively identical to human-authored changes. Similarly, there exist many accessibility tools, like speech-to-text and text-to-speech, are invaluable to those who need them *and* generally using LLMs to do so. Any potential policy should ensure that we allow accessibility tools and focus on LLM usage that creates a potential burden for maintainers, rather than focusing on LLM usage to achieve an "untainted" code base. (This is the sacrifice of "purity" described in the summary.)

> Ultimately, the goal is to avoid a situation where users are encouraged to be dishonest about LLM usage, since this creates a situation where everyone is uncomfortable. Many LLM users, including team members, have indicated that they might simply continue using LLMs and avoid disclosure for fear of repercussions, and this is a very uncomfortable position to be in. It means that LLM users are encouraged to be dishonest about their actions, and it means that maintainers are forced to accuse users of LLM usage whenever they're suspicious, which can be hostile. This is combined on top of the mention of *trivial* LLM usage, as mentioned before: if we don't distinguish usage that actually affects the end result, people stop caring and we stop knowing whether the result is affected, which makes reviewing contributions difficult.

Reviewer 24: Slightly nitpicky but I think the opening here would read better and clearer without the double negatives:

> Ultimately, the goal is to avoid a situation where users are encouraged to be dishonest about LLM usage, since this creates a situation where everyone is uncomfortable.

might become:

> Ultimately, the goal is to encourage users to be honest about LLM usage, since this promotes an environment of collaboration in good faith.

This rephrasing would also reduce word count (slightly) and the good faith phrasing would tie in with this wording later in the paragraph too:

> ... LLM users are encouraged to be dishonest about their actions, and it means that maintainers are forced to accuse users of LLM usage whenever they're suspicious ...

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: I like this proposal; if you hadn't given the example I probably would have struggled to word it, but when I get to it I'll update that line and make sure that there aren't any others that should be updated too.

Even if we are ultimately intending for the double negative, wording more positively is good IMHO.

Author: Version that I will probably settle on for now:

> Ultimately, the goal is to encourage users to be honest about LLM usage, since this promotes an environment of collaboration in good faith. Many LLM users, including team members, have indicated that they might simply continue using LLMs and avoid disclosure for fear of repercussions, and this is a very uncomfortable position to be in. It means that LLM users are encouraged to be dishonest about their actions, and it means that maintainers are forced to accuse users of LLM usage whenever they're suspicious, which really doesn't feel like good-faith collaboration. This is combined on top of the mention of *trivial* LLM usage, as mentioned before: if we don't distinguish usage that actually affects the end result, people stop caring and we stop knowing whether the result is affected, which makes reviewing contributions difficult.

Going to mark this as resolved, even though this version hasn't been published yet. Still working on it.

Note that a lot of these concerns are fundamentally fuzzy; it's in the best interest of the AI industry to promote the positive aspects of their technology while hiding the negative aspects, and so, a lot of data is intentionally unavailable. Similarly, there are a few *extremely* bad actors which do not necessarily represent the industry as a whole: for example, [xAI's data center in Memphis is explicitly polluting the air of the nearby, historically Black community][xAI Memphis], and most of the proposed/built data centers are not *this* bad.

[Rust Project Perspectives]: https://rust-lang.github.io/perspectives-on-llms/index.html
[xAI Memphis]: https://time.com/7308925/elon-musk-memphis-ai-data-center/

> Due to the extreme rift between the various arguments for and against LLMs, this section has the potential to be discussed in a generally uncritical way. Please *do not* attempt to refute or reinforce these arguments in the RFC discussion. As usual, constructive revision of wording and addition of sources is encouraged and helpful, but nonconstructive critique is unhelpful. We strongly encourage you to read the full RFC before commenting on these sections.

Reviewer 3: I think this is a reasonable request if these points are meant to illustrate a spectrum of the potential concerns people have, but at the moment several of these are written as statements of fact, many of which are objectively false, and the caveats spelled out above make it seem like the RFC accepts the facts presented but doesn't necessary endorse the ethical conclusion, whereas I *believe* the intent is to simply relay what has been said without making any statement about its truth.

Maybe these could be put in block quotes or re-worded to make it clear that these are opinions that have been expressed.

Some examples:

>  To one extent or another, *Large* Language Models will include data that was not taken with permission, i.e. stolen.

Whether or not the project (or the law...) considers this stealing seems like it's a point the RFC is not intending to settle one way or the other.

> The lack of thorough manual review for data leads to a number of issues in the output that will only become more difficult to fix as models increase in size.

I don't think there's enough evidence to settle this either way.

> Additionally, it's worth clarifying that LLMs are fundamentally *interpolary* systems, not *extrapolary* systems,

There is no definition for "interpolary" vs "extrapolary" which an LLM fails but a human passes. Can LLMs recognise patterns and extrapolate beyond the data-set: this has been proven many times. Can LLMs make accurate predictions about something outside of the data-set when no evidence of the pattern exists within the data-set? No, that would be magic even for a human.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: So, to respond point-by-point:

> at the moment several of these are written as statements of fact, many of which are objectively false, and the caveats spelled out above make it seem like the RFC accepts the facts presented but doesn't necessary endorse the ethical conclusion, whereas I believe the intent is to simply relay what has been said without making any statement about its truth.

If you have explicit examples of opinions stated as fact, please feel free to point them out. Otherwise, this feedback is unhelpful.

> > To one extent or another, Large Language Models will include data that was not taken with permission, i.e. stolen.
>
> Whether or not the project (or the law...) considers this stealing seems like it's a point the RFC is not intending to settle one way or the other.

Taking something without permission is stealing, and not all stealing is considered bad to the point of a crime. For example, taking a pen from someone's desk can be called "stealing" but that doesn't mean you deserve to be arrested for it.

> > The lack of thorough manual review for data leads to a number of issues in the output that will only become more difficult to fix as models increase in size.
>
> I don't think there's enough evidence to settle this either way.

These models have training sets that comprise effectively the size of the entire internet. Do you think that someone personally read through every piece of training data? Do you think that even a small army of underpaid Kenyans read it, and that this would constitute "thorough review"? This is not a claim that requires evidence; the scale of these models, if it were even a fraction of what's claimed, prevents you from thoroughly reviewing all the data. And, since the main mechanism for improving models is adding more data, this only gets worse with time.

> > Additionally, it's worth clarifying that LLMs are fundamentally interpolary systems, not extrapolary systems,
>
> There is no definition for "interpolary" vs "extrapolary" which an LLM fails but a human passes. Can LLMs recognise patterns and extrapolate beyond the data-set: this has been proven many times. Can LLMs make accurate predictions about something outside of the data-set when no evidence of the pattern exists within the data-set? No, that would be magic even for a human.

This isn't the point being made here. The point being made is that without logical deduction, LLMs are much more constrained by their input data than humans are. This is also a quick summary to reference a later point in the section, which includes an explicit example of what this entails.

I'd be happy to reword this to be more clear, but I'm not sure whether this is actually unclear or if you just dislike this point.

Reviewer 3: > If you have explicit examples of opinions stated as fact, please feel free to point them out. Otherwise, this feedback is unhelpful.

I gave some examples? I didn't try to be exhaustive because I didn't think the discussion was meant to be about whether or not they were true or not, since you said:

> Please *do not* attempt to refute or reinforce these arguments in the RFC discussion.

Yet seemingly you do want to argue these points?

> Taking something without permission is stealing, and not all stealing is considered bad to the point of a crime. For example, taking a pen from someone's desk can be called "stealing" but that doesn't mean you deserve to be arrested for it.

Either these statements are part of the "RFC proper" and should be open to discussion or they are not, there's no point discussing them and they shouldn't be presented as such.

Since you do apparently want to discuss them though, here goes:

> This is not a claim that requires evidence; the scale of these models, if it were even a fraction of what's claimed, prevents you from thoroughly reviewing all the data. And, since the main mechanism for improving models is adding more data, this only gets worse with time.

It is already the case that the vast majority of data does not go though any kind of manual review process at all, so this is irrelevant. You'd have to somehow evidence that the increase in data doesn't make up for the percentage-wise decrease in manual reviews when it comes to LLM quality, and there's no way to answer that at the moment.

> This isn't the point being made here. The point being made is that without logical deduction, LLMs are much more constrained by their input data than humans are.

a) That's not what "interpolary" means, so if that's the argument you're trying to make I don't understand why you would use those words.

b) Humans are notoriously bad at logical deduction unless they resort to formal methods. There's a reason we invented systems like propositional logic. LLMs have access to the same tools. I'm sure there's a quantitative difference in ability here (for now) but I don't believe anyone can say with certainty whether there is a qualititative difference. If you can point to a paper on this I would be very interested to read it.

If you're trying to argue that LLMs empirically seem to be worse at certain things... just say that? It weakens the argument to be overly (and incorrectly) specific about *why* something is the case, when all that matters is that it appears to be the case.

> I'd be happy to reword this to be more clear, but I'm not sure whether this is actually unclear or if you just dislike this point.

To be honest I think the guide and reference sections of this RFC are fine. It's a good compromise that I hope will ensure a high quality of contributions going forward.

The rest of the RFC reads like an essay on your personal beliefs on LLMs, some of which I think are substantiated, some not. I think it only harms the ability to come to a consensus to include them as anything other than "concerns that have been raised".

Author: > > If you have explicit examples of opinions stated as fact, please feel free to point them out. Otherwise, this feedback is unhelpful.
> 
> I gave some examples? I didn't try to be exhaustive because I didn't think the discussion was meant to be about whether or not they were true or not

To be clear, I… genuinely did not even think these were examples here. Because to me, these read entirely as reasonable deductions to make based upon the limited things we know about LLMs.

> > Please do not attempt to refute or reinforce these arguments in the RFC discussion.
>
> Yet seemingly you do want to argue these points?

I agree this is a bit confusing; what I meant to say here is that I don't want to accept feedback of the form "hey, here's another example of ____" or "I disagree that LLMs are unethical; this argument is unsatisfactory" since I don't think they're going to be very productive. But these points, to me, *are* just reasonable statements being deduced, and I'm responding to them because I want to understand your viewpoint here.

> It is already the case that the vast majority of data does not go though any kind of manual review process at all, so this is irrelevant. You'd have to somehow evidence that the increase in data doesn't make up for the percentage-wise decrease in manual reviews when it comes to LLM quality, and there's no way to answer that at the moment.

So, to me, it feels like a given that model creators are just going to keep adding more data to their models, *and* that they will be less and less able to thoroughly review all that data as a result, and the more debatable point is whether the thorough review is necessary. What it felt like you were saying is that more evidence is needed to prove that thorough review is not happening when like… no, this is just not possible; most of it is not reviewed at all, even.

That said, I don't try to point this out as fact; it's specifically referencing a later point where I argue that bias in the models is a side effect of the lack of review. This definitely is a logical leap but I think it's a fair one.

> The rest of the RFC reads like an essay on your personal beliefs on LLMs, some of which I think are substantiated, some not. I think it only harms the ability to come to a consensus to include them as anything other than "concerns that have been raised".

An RFC itself is an essay that reflects one's opinions. If proposals were an obvious choice that didn't involve any opinions we wouldn't have the RFC process.

I tried my best to make sure the arguments here were as clear as possible and motivate using evidence instead of opinions wherever possible. There is an inherent bias in the motivation because, well, as I said, a policy proposal is an inherently biased position; I'm biased toward the proposal being accepted. I'm intentionally overloading on arguments because I feel like a lot of people haven't been listening to any arguments, but if some arguments are confusing, I do want to try and clarify them.

----

The last bit on "interpolary" versus "extrapolary", I agree that the choice of words is probably bad, and the section it's leading into also needs some work too, as demonstrated by this thread: https://github.com/rust-lang/rfcs/pull/3959#discussion_r3186291305

Like that thread, I'm also going to take some time to think about the best way to reword it.

To respond to this, though:

> b) Humans are notoriously bad at logical deduction unless they resort to formal methods. There's a reason we invented systems like propositional logic. LLMs have access to the same tools. I'm sure there's a quantitative difference in ability here (for now) but I don't believe anyone can say with certainty whether there is a qualititative difference. If you can point to a paper on this I would be very interested to read it.

Just to make sure we're clear, what I think you're proposing is the idea of an LLM running some code to perform logical deduction to perform the equivalent of what a human "doing deduction manually" might be. We could extend the problem also to performing basic arithmetic as well.

As far as I'm aware, no model successfully does this correctly, and doing so does not avoid any issues with bias based upon symbolic information (i.e. GSM-Symbolic still fails), but my information could be out of date.

Reviewer 13: > Just to make sure we're clear, what I think you're proposing is the idea of an LLM running some code to perform logical deduction to perform the equivalent of what a human "doing deduction manually" might be. We could extend the problem also to performing basic arithmetic as well.

> As far as I'm aware, no model successfully does this correctly, and doing so does not avoid any issues with bias based upon symbolic information (i.e. GSM-Symbolic still fails), but my information could be out of date.

Well, if you look at [AI contributions to Erdős problems](https://github.com/teorth/erdosproblems/wiki/AI-contributions-to-Erd%C5%91s-problems)…

Many of the "AI standalone" solutions involve the LLM writing formal proofs in Lean, which sounds pretty close to what you said.  I doubt this technique would be very applicable to Rust though.

Among the proofs that did not use Lean, many are marked incorrect, suggesting that even in mathematics (which LLMs have shown to be relatively good at), hallucinations are still an issue.  Human mathematicians make mistakes too, but probably not as many, and/or they're better at catching them on review.

Still, the fact that LLMs have moved up from grade-school word problems (as seen in the GSM-Symbolic paper) to unsolved mathematical problems tends to put those issues in perspective.  They are much better at logical reasoning than they used to be.  

(Personally I have similar "losing the will to live speedrun" feelings about LLMs as mentioned in the RFC, but not when I see them write bad code.  I have those feelings when I see them write good code.  I agree that programming is an incredibly fun task, but it is also a useful task, and I fear that it will cease to be so.)

Author: So, FWIW, I have been following Terry Tao's journey through talking about the Erdős problems and using various tools to solve them and still don't know how I feel about it. But, like I mentioned, I don't think that these models are capable of actually *reasoning* with the help of Lean or some other tool, just writing proofs as an end goal.

Reviewer 25: Tao has some astute remarks at 1h45m here:
https://www.youtube.com/live/tN4hsT5t0nw?si=cIQj2Di6sNdZHr7P&t=6330

AI have been used in different ways by mathematicians:  Individual mathematicians have conversations with LLMs, which can range from literature search to using the LLM as an editor to seeing how far they can push the LLM.  Proof writing in Lean, sometimes with Lean giving the AI fast feedback.  

> b) Humans are notoriously bad at logical deduction unless they resort to formal methods. There's a reason we invented systems like propositional logic. 

Agreed largely.  We humans take ages to learn formal methods though.

> LLMs have access to the same tools. 

Afaik not really.  Google's search AI gives the wrong units all the time.  3+ years ago Google avoided this by using verbatim quotations, which LLMs made worse.

I do think fast feedback between LLMs and real formal tools shows promise here, both in massively reducing the LLM size & cost, and in results that look closer to actual reasoning.  At the fancy end like Lean, one feels like Lean deserves more credit than the LLM that wrote the Lean program.  Yet, this principle applies even for easier stuff:  Why should LLMs do things for which better tools exist?  Why should LLMs parse code when they could use tree-sitter?  Why should LLMs ever do computations when they could invoke some units aware calculator?  etc  And LLMs maybe more useful in Rust once you go beyond the type system, like fuzzers and static analysis.  That said..

Tao's talk focused upon the problem of learning anything from the LLM proofs, because LLMs really suck at elegance.  We've buggy sandboxing everywhere like bad AppArmor and SE Linux profiles.  You'd think LLMs could help here, especially if examined the source code.  Yet, they need some tools to more elegantly communicate what's unique, what context and assumptions apply, etc.  It's still the elegance problem.

Author: I think this is maybe getting a bit off topic, although to be fair, I kind of invited the discussion by adding my own remarks. The main point I was trying to make was that Tao's work doesn't really change what's already been said, which is that these tools don't reason very well on their own, and this section also needs to be reworded a bit anyway.

Author: I think this is far enough off topic I'm just going to resolve this. Ultimately, I don't think that mentioning Tao's work is helpful for the RFC, and the original comment appears to have been written in bad faith.

The Rust Foundation, located in the U.S., has consulted its own legal counsel on the matter of whether the project should be concerned about copyrightability of LLM output. The relayed report is as follows:

> > On [the matter of copyright of LLM generated contributions], the Rust project directors consulted the Rust Foundation's legal counsel and they did not have significant concerns about Rust accepting LLM-generated code from a legal perspective. Some courts have found that AI-generated code is not subject to copyright and it's expected that others will follow suit. Any human-contributed original expression would be owned by the human author, but if that author is the contributor (or the modifications are licensed under an open source license), the situation is no different from any human-origin contribution. However, this does not present a legal obstacle to us redistributing the code, because, as this code is not copyrighted, it can be freely redistributed. Further, while it is possible for LLMs to generate code (especially small portions) that is identical to code in the training data, outstanding litigation has not revealed that this is a significant issue, and often such portions are too small or contain such limited originality that they may not qualify for copyright protection.

Reviewer 15: maybe we could link to the source here https://rust-lang.github.io/perspectives-on-llms/feb27-summary.html#the-legality-of-ai-usage.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: So, I wasn't 100% sure if that actually was the source or not, since it was also shared on Zulip and I'm not 100% sure where it was shared first. But I can include that if it is actually the primary place it should be linked.

LLMs, largely, have completely disregarded that respect. Respect would be only using things that you've explicitly gotten with permission, which is explicitly not what they have done. For example, this was listed as evidence in [Kadley v. Meta](https://www.courtlistener.com/docket/67569326/417/1/kadrey-v-meta-platforms-inc/):

> This document appears to be notes from a January 2023 meeting that Mark Zuckerberg attended. It is heavily redacted, including a large section titled "Legal Escalations." Immediately after that section the document states "[Zuckerberg] wants to move this stuff forward," and "we need to find a way to unblock all this."

Several people were attempting to find a way to properly obtain licenses for copyrighted material before proceeding. Then, suddenly, the CEO of the company demonstrates his desire to "move this stuff forward," and people just start doing it without permission. Even if the employees responsible for creating the model said "using pirated material should be beyond our ethical threshold," the CEO decided to ignore those concerns. Even though Meta's LLM is not a coding model, their case is not particularly unusual in the industry.

> And, it's worth mentioning that the "worst case" scenario of xAI, brought up earlier, *is* supported by GitHub Copilot, showing that at least all of the "good actors" in the AI space are willing to work with all the bad actors on equal footing. This example indicates that Meta's case is likely to be the norm.

Reviewer 16: ```suggestion
And, it's worth mentioning that the "worst case" scenario of xAI, brought up earlier, *is* supported by GitHub Copilot, showing that at least many of the "good actors" in the AI space are willing to work with many of the bad actors on equal footing. This example indicates that Meta's case is likely to be the norm.
```

I think there are likely a few exceptions, so "all" is overbroad.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..817509d7a18f43476d162df25a7c9a5a048902aa)*

Author: I think the second "all" / "many of" is actually redundant here, but yes, I can reword this.

Author: So, I am unresolving this and heavily considering at least strengthening the wording here given: https://www.cnbc.com/2026/05/06/anthropic-spacex-data-center-capacity.html

Now that Anthropic is explicitly using the data centre in question, I think it's unreasonable to not point out how actually bad this case is, even if I do concede it's not *literally* everybody.

Author: To be clear, this is now:

> And even allegedly "ethical" actors in the AI space are more than willing to compromise on ethics for the sake of their models. One obvious example is how GitHub Copilot supports Grok, run by xAI and their horrendous Colossus data centre, but a more recent example is Anthropic deciding to outright use said data centre to power Claude Pro. As the fight to the bottom continues, these companies are more and more willing to "compromise" on their values, and we should not wait to see how low they can go.

A more noticeable change comes is semiconductor technology, particularly silicon wafers. While silicon itself is extremely prevalent ("it's just sand"), the purification of silicon wafers for producing semiconductors is very costly, and the AI industry has been allocating more and more wafer output for costly, lower-yield technologies like HBM (high-bandwidth memory) and stacked NAND storage. This results in a noticeable decrease in wafers that can be used for the technologies that get used on consumer devices, and an increase in the price of the specific technologies used for AI (memory and storage).

While the cryptocurrency rush of the past decade resulted in increased GPU prices, the AI industry has increased prices for a number of semiconductor components across the board, particularly DRAM (memory) and NAND flash (storage). The below graphs from [pcpartpicker.com] indicate trends in pricing of a select few components between late 2024 and early 2026. Note that these costs are not for raw components, but the end products that users might purchase to build a computer.

<!-- dear GitHub, you should actually support <url> syntax properly instead of going all-in on LLMs -->

> [pcpartpicker.com]: https://pcpartpicker.com

Reviewer 15: There's nothing wrong with GitHub here, by CommonMark specification a [URI autolink](https://github.github.com/gfm/#autolink) requires an "absolute URI" which must includes a `:`[^1]. The behavior is the same in rustdoc and pandoc. If you don't want a reference-style link you should include the `https://` part like:

```suggestion
While the cryptocurrency rush of the past decade resulted in increased GPU prices, the AI industry has increased prices for a number of semiconductor components across the board, particularly DRAM (memory) and NAND flash (storage). The below graphs from <https://pcpartpicker.com> indicate trends in pricing of a select few components between late 2024 and early 2026. Note that these costs are not for raw components, but the end products that users might purchase to build a computer.
```

[^1]: to be precise, an absolute URI must match the regex `[a-z][-.+0-9a-z]{1,31}:[^\s\p{C}<>]*`

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: Huh, TIL. I'll remove the comment but since the protocol makes it a bit annoying to read, I'd prefer to keep it out of the text of the paragraph.

One load-bearing aspect of these models is that the highest-weighted path is not necessarily the most correct one, and while it is possible to simply take an alternative path from the same starting state, this is not the choice that models take. Instead, models tend to create a feedback loop where some external information about the path and its potential faults is fed back into the model as new context. With this new context, hopefully, the model will be more likely to weight the correct path higher on the list, and if it doesn't, well. Since this loop is technically unbounded, that's great when your billing model depends on the number of used tokens.

#### The Brute-force Loop

> Claude Code, often cited as one of the best models, has a great example in its source code which it accidentally leaked. The below code samples are only used for educational purposes and cannot be used alone to create a commercial competitor to Claude Code, and so, do not infringe on Anthropic's copyright.

Reviewer 15: ```suggestion
Claude Code, often cited as one of the best agentic coding tools, has a great example in its source code which it accidentally leaked. The below code samples are only used for educational purposes and cannot be used alone to create a commercial competitor to Claude Code, and so, do not infringe on Anthropic's copyright.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: This feels pedantic to me, but I would love to know why you think this is better. It's both a model and a tool, but in this case it *is* going through the source for the tool, not the model.

Reviewer 15: The model is the LLM itself, like Claude Opus 4.7 / GPT 5.5 / etc. The tool (Claude Code / Codex / etc) invokes the model and interprets the output to perform commands. Claude Code is definitely not a model.

You may think of the current sentence being like someone saying "***Cargo**, often cited as one of the best **programming languages**, …*".

Author: Yeah, in hindsight, I did kind of answer my own question.

},
```

Ultimately, the relevant code that runs these tools is located here:

```ts
async function checkPermissionsAndCallTool(
  tool: Tool,
  toolUseID: string,
  input: { [key: string]: boolean | string | number },
  toolUseContext: ToolUseContext,
  canUseTool: CanUseToolFn,
  assistantMessage: AssistantMessage,
  messageId: string,
  requestId: string | undefined,
  mcpServerType: McpServerType,
  mcpServerBaseUrl: ReturnType<typeof getLoggingSafeMcpBaseUrl>,
  onToolProgress: (
    progress: ToolProgress<ToolProgressData> | ProgressMessage<HookProgress>,
  ) => void,
): Promise<MessageUpdateLazy[]> {
  // Validate input types with zod (surprisingly, the model is not great at generating valid input)
  const parsedInput = tool.inputSchema.safeParse(input)
  if (!parsedInput.success) {
```

Ultimately, this over-1200-line TypeScript function performs the standard loop of Claude Code:

1. Input is massaged into the correct format to send to the model
2. The model runs the resulting prompt
3. Based upon the output, it may run some code or external tool
4. Based upon the results of synthetic tools, it may provide additional feedback to the model

> 5. All of this is collected and shunted back into the loop

Reviewer 15: Sorry I find it impossible to follow what you're trying to say in this section.

The snippet at lines 234-258 does not contain `TelemetrySafeError_…`, despite it opens with "let's consider `TelemetrySafeError_…`" in line 232. Also `AnalyticsMetadata_…` are seen being cast from `number | boolean | string` so the type does not look like a "subclass" of `TelemetrySafeError_…`. 

Lines 260-264 mentioned JSON validation, but none of the snippets here contain any code that runs the `call()` function in lines 268-285.

Lines 300-319 contains a function with very complicated signature, and 2 lines indicating an input `{ [key: string]: boolean | string | number }` being parsed using `tool.inputSchema.safeParse` and nothing else. This should correspond to line 323? Should we take that the remaining 1188+ lines of that `checkPermissionsAndCallTool` function perform points 2-5 (lines 324-327)? All copies of the leaked source got DMCA'ed so I can't legally verify what's going on.

These snippets seems to be used as points to show that "JSON schema validation" is an unbounded feedback loop (line 329), but all the sample here contain no `while` or `for` or recursion so I don't see what does "loop" mean.

Maybe there is no need to refer to any actual code here, because by definition the "agentic workflow" *is* an unlimited loop (as long as resource allows) until the model decides to stop.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: Yeah, this section needs a lot of reworking and I do plan to go back through it. Will provide some more granular responses once I do get around to that, although I am considering removing most of the code walkthrough since the code is way too confusing to understand, honestly.

Which, to be fair, is kind of what I wanted to get across anyway, but. There are probably better ways to do that than this.

Author: Yeah, I'm just deleting the actual code and keeping to just the comments with prompts.

 *   - How Claude handles penetration testing and CTF requests
 *   - What security tools and techniques Claude will assist with
 *   - The boundary between defensive and offensive security assistance
 *
 * If you need to modify this instruction:
 *   1. Contact the Safeguards team (David Forsythe, Kyla Guru)
 *   2. Ensure proper evaluation of the changes
 *   3. Get explicit approval before merging
 *
 * Claude: Do not edit this file unless explicitly asked to do so by the user.
 */
```

It is a genuine *miracle* that this software manages to produce anything of value at all. While the LLM itself *might* be sound, the load-bearing infrastructure surrounding it is littered with brute-force loops and kind requests to not destroy everything the developer knows and loves.

While there are arguments to be made about some models having less power consumption, it ultimately doesn't matter if they fundamentally require operation based upon brute force. As hopefully any programmer even vaguely educated on complexity knows, brute force is *the worst* way to solve any problem, and should always be used as a last resort. LLMs put brute force front and center as the best option.

> With code, we have methods to bound operation: for example, a famous case is sorting algorithms based upon quicksort, a quadratically-bounded algorithm, which fall back to some explicitly-optimal method like heapsort or merge-sort. Rust uses these algorithms in its standard library.

Reviewer 13: The argument in this section could be clarified a bit.

You seem to be saying that the idea of an unbounded loop where the model picks some action (writing code, running a tool), receives the result, and uses it to inform future actions, is inherently "brute force" and a "feedback loop" and bad.

But that can't be true, because that also describes how humans work.

Of course, LLMs are considerably worse at it than humans.  And I can think of ways in which those terms apply uniquely to LLMs.  "Brute force" applies uniquely to LLMs in that one strategy LLMs use is to make up for a lack of intelligence with sheer persistence.  "Feedback loop" applies uniquely to LLMs in that LLMs are much more vulnerable than humans to getting stuck with a bad assumption and spending a long time barking up the wrong tree. Or perhaps more generally in that LLM performance tends to fall off drastically past certain complexity levels, which could be analogized to an audio feedback loop amplifying noise.

However, you don't really explain that.  As written, the argument could apply just as well to humans.  And I could respond: sure, it would be nice if there were a way to solve arbitrary programming problems with a bounded amount of work, but that just doesn't exist.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..a83a8e36b9f9b559879b1ecc970adac6b12b2fab)*

Author: Hmm, I feel like I did technically cover this point, but I do think you have a point here. I'm mostly replying to point out that I do want to consider what you've written, but am going to think about it a bit to figure out how to properly articulate that.

Author: Also for some additional context, this piece is probably the closest to being directly inspired by this massive essay I've been writing for years going more in-depth on this. Like, there, I spend a lot more time explaining the specifics of complexity theory, incompleteness proofs, and other things to make a point, but like, I both haven't written that fully and am not going to write that fully here. But I feel like there's some nugget of information that could be gleaned without going past the surface level.

Author: So, this section has been reworked a bit; let me know what you think.

Author: Gonna mark this as resolved for now just for clutter reasons since things have changed a lot, but please feel free to unresolve or open a new thread if you have additional stuff you want to point out!

This leads into the second main part of the policy, which is *disclosure*. *Nontrivial* LLM usage should always be disclosed; we don't have any standard format for this and simply ask you explain in your issue, PR, etc. that an LLM was used, and ideally how. Similarly to how explaining your general thought process can be helpful for reviewing changes, explaining the tools you used and how can help people understand what they're looking at and identify potential quirks. Disclosure should also be included in the descriptions for PRs; a commit message header is not sufficient.

If LLM usage falls in the gray area of "research," then disclosure is only requested if a maintainer is confused or asks what your process was. In general, this is the preferred, non-accusatory way of requesting more details about a contribution: "what was your thought process when writing this?" instead of "did you use an LLM for this?"

### Slop is Strictly Moderated

> Contributors are expected to put in the effort to fully understand their changes, and this means both validating any research and ensuring that any LLM-authored code is accurate. A particular case of this not happening, called *slop*, occurs when an author appears to have both used an LLM to create a change and done very little work of their own to verify the result. If you're worrying your work might be considered slop, you probably *already* haven't qualified, because simply worrying about it usually implies that you've put in at least a little effort.

Reviewer 15: ```suggestion
Contributors are expected to put in the effort to fully understand their changes, and this means both validating any research and ensuring that any LLM-authored code is accurate. A particular case of this not happening, called *slop*, occurs when an author appears to have both used an LLM to create a change and done very little work of their own to verify the result. If you're worrying your work might be considered slop, the work probably *already* isn't sloppy, because simply worrying about it usually implies that you've put in at least a little effort.
```

OK I got the explanation why I find the original expression in the first read so strange. I suppose this sentence is telling 

* 1 - you *worried* the work might be slop
* 3 - *worrying* implies you have "put in at least a little effort"
* 2 - *putting effort* means the work is not slop

but previously the verb "Qualified" was used here, and "Qualified" is typically used for good things, so "Qualified [slop]" sounds oxymoron. When treating "Qualified" to mean something good you should do, the clauses 1+2' makes it sounds like "Simply worrying is *not* enough to make the work not-slop", but then contradicts with the clause 3.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: Yeah, okay, this is just better wording. I do think that we should use "slop" here instead of "sloppy" since, even though it is just a conjugated form, we should try to be as consistent with wording to make it clear that we mean "LLM slop" here, and not just, the code being generally sloppy.

Author: So, new version:

> Contributors are expected to put in the effort to fully understand their changes, and this means both validating any research and ensuring that any LLM-authored code is accurate. A particular case of this not happening, called *slop*, occurs when an author appears to have both used an LLM to create a change and done very little work of their own to verify the result. If you're worrying your work might be considered slop, you probably *already* didn't meet the criteria for being slop, because simply worrying about it usually implies that you've put in at least a little effort.

Reviewer 15: ok better, though I'd still change the subject as "***your work** probably already didn't meet the criteria for being slop*", since "slop" is used to describe the product, not the person.

#### Team-Specific Usage

For projects which are entirely owned by one team (e.g. cargo, rustfmt, rust-analyzer, etc.), teams are allowed to fully control their own policies regarding LLM usage. This means that, at the moment, teams are completely unrestricted regarding LLM usage as long as it follows this policy. That means a few things:

* Teams are allowed to add custom model configurations (like `AGENTS.md` or `CLAUDE.md`) to the repositories if they desire.
* Teams may elect to allow LLM-based review tools or issue bots.
* Teams may elect to allow LLM-authored issues and PRs as long as they are not *slop*.
* Teams are encouraged to clarify via PR, issue templates, etc. what their policies are.

This policy may change in the future to ensure consistency across the project.

#### Shared and Public Usage

Some repositories, particularly `rust-lang/rust` and `rust-lang/rfcs`, are shared across teams, and cannot easily have team-specific policies. Additionally, public communications from the project affect the reputation of the project and should have a high standard for quality.

As a result, certain LLM usage is restricted for shared repositories in certain instances:

> * Model-specific configurations should not be included in repositories. Including them in `.gitignore` to allow user-specific configurations is fine.

Reviewer 15: ```suggestion
* Model-specific configurations should not be included in repositories, preferably excluded via `.gitignore`. Private use of user-specific configurations (like `AGENTS.override.md` and `CLAUDE.local.md`) is fine.
```

The original sentence sounds like "copying the LLM config verbatim into `.gitignore` is fine" which I have to parse many times to understand 🤷.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: Reworded as:

> Model-specific configurations should not be included in repositories. Some of the files involved may be mentioned in `.gitignore`.

#### Decisionmaking

Since there is a potential for bias in models, in general, the "final decision" on any action should come from the conscious decision of a team member, not an LLM. This also includes "filtering" cases where a set of options is narrowed down, e.g. a list of potential grant nominees or features to be implemented. Ultimately, human team members should be making the decision here, not LLMs, although this policy does not forbid team members from using LLMs for advice.

As a particular example of this policy, if an LLM is used for review in a team-specific repository, that review must not be binding, and a team member should have the final say.

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

### Code of Conduct changes

This RFC proposes replacing the following line:

* Please keep unstructured critique to a minimum. If you have solid ideas you want to experiment with, make a fork and see how it works.

With the below lines:

* Reviewing changes takes effort, and you should be mindful to avoid adding more work for maintainers. In general, you should understand all changes you make and be willing to explain them.

> * Whenever someone asks questions, assume good faith and respond honestly. In order to effectively work together, we need to know what we're working on.

Reviewer 15: "Replacing" as in "Deleting the item in line 530 and Add the items in lines 534-535"? I don't see how the line 530 item is related to LLM and needs to be deleted 🤔 

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..68e00bde3e72ed51a74775c8c7a974f19ee00b77)*

Author: Yes, replacing. It's not necessarily related to LLMs, but my thought process is that the proposed two items effectively include the former in a bit more of a clear way, which is why I decided to do it this way.

That said, if you do disagree, I could probably just make it an addition instead of a replacement.

Author: Added a note clarifying this.

* 2026 Mar 23: Jack and Oli [merge the retracted blog post](https://blog.rust-lang.org/2026/03/20/rust-challenges/)
* 2026 Mar 24: Jack [weighs in on the discussion](https://rust-lang.zulipchat.com/#narrow/channel/392734-council/topic/Vision.20Doc.20blog.20post.20and.20LLM.20usage/near/581387730)
* 2026 Mar 28: Jakub opens an RFC draft: [Add policy for using AI in official Rust Project communication channels](https://github.com/Kobzol/rfcs/pull/1)
* 2026 Apr 09: Oli [ends the discussion moratorium](https://rust-lang.zulipchat.com/#narrow/channel/392734-council/topic/Vision.20Doc.20blog.20post.20and.20LLM.20usage/near/584369859)

Before Jack's post, Jieyou Xu offered a revised version of the compiler MCP to the leadership council to adopt as a project-wide policy, which experienced several versions of revision:

* 2026 Mar 06: Jieyou Xu (@jieyouxu) opens a leadership council issue: [Policy proposal: No low-effort contributions](https://github.com/rust-lang/leadership-council/issues/273)

> * 2020 Mar 20: (Jack's post happens here)

Reviewer 26: ```suggestion
* 2026 Mar 20: (Jack's post happens here)
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..a83a8e36b9f9b559879b1ecc970adac6b12b2fab)*

The second is the data itself, both in content and the means by which that content is filtered. Specifically, the *Large*ness of LLMs requires an amount of data past the point where thorough manual review is possible, and this only becomes a bigger problem as more data is used. The lack of thorough manual review for data leads to a number of issues in the output that will only become more difficult to fix as models increase in size.

> Additionally, it's worth clarifying that LLMs' fundamentally are more sensitive to the data used to create them than other modeling methods. Since they have no inherent capacity to perform reasoning, they rely heavily on the contextual relation between terms to achieve accuracy. This means that a lot of topics that are discussed less in general, which therefore show up less in the source data, are substantially less accurate in LLM output.

Reviewer 14: ```suggestion
Additionally, it's worth clarifying that LLMs fundamentally are more sensitive to the data used to create them than other modeling methods. Since they have no inherent capacity to perform reasoning, they rely heavily on the contextual relation between terms to achieve accuracy. This means that a lot of topics that are discussed less in general, which therefore show up less in the source data, are substantially less accurate in LLM output.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..f47c2634f021e67ea4399e42d9934304a4644257)*

One of the most fundamental parts of problem solving is breaking problems into smaller, more manageable chunks. Effectively, since brute force has exponential complexity, you want to ensure that the exponent is as small as possible if you can't get rid of it. And, well, *reality* has this sort of exponential complexity that shows that LLMs fundamentally won't be able to solve problems if you encounter something they've never seen before.

> And, unlike humans, the additional context may not actually help these tools get out of the feedback loop. If the answer to a problem ultimately isn't within the bounds of the LLM's operation, it might get stuck forever, or it might give an invalid solution just to end the loop. There are *so many* examples of problems that LLMs fail spectacularly at, and while many of them could be described as "trick questions"… those exist in the real world, too! Sometimes you just have a false misconception about something that leads to you begging the wrong question. So, no, it's not particularly unusual to ask whether you should walk to the car wash if it's nearby, or drive.

Reviewer 24: This section could be supported further with some citations if desirable:

- [Modarressi, A., Deilamsalehy, H., Dernoncourt, F., Bui, T., Rossi, R.A., Yoon, S., Schütze, H., 2025. NoLiMa: Long-Context Evaluation Beyond Literal Matching.
](https://doi.org/10.48550/arXiv.2502.05167)
- [Hong, K., Troynikov, A., Huber, J., 2025. Context Rot: How Increasing Input Tokens Impacts LLM Performance](https://www.trychroma.com/research/context-rot)
- [Liu, N.F., Lin, K., Hewitt, J., Paranjape, A., Bevilacqua, M., Petroni, F., Liang, P., 2023. Lost in the Middle: How Language Models Use Long Contexts.](https://doi.org/10.48550/arXiv.2307.03172) (Rather old at this point, but at least demonstrates the problem is long-standing.)

To summarize the findings: additional context can in fact degrade the performance of these tools, at least for some tasks. In essence, as the context window grows attention must be distributed across more tokens; essentially diminishing signal-to-noise ratio.

The models evaluated in these articles have since been superseded, but considering that newer models are built on the same fundamental methods (transformers, attention) I'd be surprised if the findings didn't apply to newer models as well.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3959/17a311afa36c353e88578e9dde44225aaa447ee1..6430511475dbb621a69e54a8b46ad247b321e076)*

Reviewer 15: @garland-culbreth  I think all these papers are talking about "[context window](https://en.wikipedia.org/wiki/Context_window)", i.e. the transformer taking 1 million tokens simultaneously as input to predict the next one ($T_{1000001} = \mathbb E[f(\\{T_1, \dotsc, T_{1000000}\\})]$). 

Line 316's use of "[context](https://en.wiktionary.org/wiki/context)" seems not about this LLM concept, but rather just in general sense of "environment", "background information" etc. Because the provided example failures like the "car wash" question is certainly unrelated to the prompt using up the context window. 

So these citations are likely irrelevant to this paragraph.

Author: Yeah, while I was particularly thinking about the context to the transformer as the method that usually accomplishes this, I wasn't specifically focusing on the context as a sequence of tokens, rather, as whatever abstract method of conveying context is for the model. In a lot of cases it's just going to be adding more token context, which is very similar to just increasing the overall size of the prompt, but that's not necessarily what's going on here.

Author: (Also just want to add that citing additional research, even if it ultimately doesn't get used in the RFC, is extremely appreciated. It's really difficult to navigate the literature since there's a lot of it that's just overt marketing and it's hard to tell which is which unless you're constantly reading papers as they come in.)

Author: (side note: hope that you're okay with me marking the root of this thread as *un*resolved, since it just makes things easier to read if the rest of the thread is open. I would like to keep this open since there aren't too many concurrent discussion threads and having the links to the research is helpful, but if you would rather resolve the whole thread, feel free)

Reviewer 24: (All good! I had indeed meant to resolve the whole thread but I couldn't figure out how to actually do that; I'm not sure if I have the option to resolve whole threads? Anyway, if the links are handy agree it's better to keep them visible! Apologies for the trouble)

Author: Ah, you should have the option to click "resolve conversation" below the final comment since you opened the thread; this option should always be available for people who have opened the thread or who made the PR it's on, but I know GitHub's UI has always been a mess.

It's all good, we're all figuring out how to do stuff together.

Reviewer 24: Well I do see the option now I look in the right place, but can't mark it resolved on my end because GitHub itself raises an error upon attempting:

<img width="565" height="113" alt="Image" src="https://github.com/user-attachments/assets/36375266-000d-4d2c-a783-c8cc6a6b0acb" />

A search turned up some discussion about this ([1](https://github.com/orgs/community/discussions/187503)) ([2](https://github.com/orgs/community/discussions/8327)). Apparently permissions or subsequent commits can cause GitHub to deny resolve requests from a reviewer.

I suppose author must still be able to though, so do feel free to mark this all as resolved and sorry for the side-track.

Author: Gonna resolve this since the section its citations would help has been removed, but again, thank you for providing additional literature on the subject. <3
